// @directus/themes/dist/index.js
import { computed as Ce, unref as ee, reactive as Ei, defineComponent as Ci, toRefs as Ni, createBlock as Ui, openBlock as Di, Teleport as Li, createTextVNode as Mi, toDisplayString as Wi } from "vue";
import { useHead as Bi } from "@unhead/vue";
import { get as Ct, merge as Tn, mapKeys as Ki } from "lodash-es";
import { defineStore as Vi, storeToRefs as Hi } from "pinia";
import Gi from "decamelize";
import { flatten as qi } from "flat";
import { cssVar as Ji } from "@directus/utils/browser";
function f(e, r, n) {
  function t(a, l) {
    var p;
    Object.defineProperty(a, "_zod", {
      value: a._zod ?? {},
      enumerable: !1
    }), (p = a._zod).traits ?? (p.traits = /* @__PURE__ */ new Set()), a._zod.traits.add(e), r(a, l);
    for (const b in c.prototype)
      b in a || Object.defineProperty(a, b, { value: c.prototype[b].bind(a) });
    a._zod.constr = c, a._zod.def = l;
  }
  const o = n?.Parent ?? Object;
  class u extends o {
  }
  Object.defineProperty(u, "name", { value: e });
  function c(a) {
    var l;
    const p = n?.Parent ? new u() : this;
    t(p, a), (l = p._zod).deferred ?? (l.deferred = []);
    for (const b of p._zod.deferred)
      b();
    return p;
  }
  return Object.defineProperty(c, "init", { value: t }), Object.defineProperty(c, Symbol.hasInstance, {
    value: (a) => n?.Parent && a instanceof n.Parent ? !0 : a?._zod?.traits?.has(e)
  }), Object.defineProperty(c, "name", { value: e }), c;
}
class rr extends Error {
  constructor() {
    super("Encountered Promise during synchronous parse. Use .parseAsync() instead.");
  }
}
class bo extends Error {
  constructor(r) {
    super(`Encountered unidirectional transform during encode: ${r}`), this.name = "ZodEncodeError";
  }
}
const Fo = {};
function Ne(e) {
  return Fo;
}
function Yi(e) {
  const r = Object.values(e).filter((t) => typeof t == "number");
  return Object.entries(e).filter(([t, o]) => r.indexOf(+t) === -1).map(([t, o]) => o);
}
function En(e, r) {
  return typeof r == "bigint" ? r.toString() : r;
}
function Vn(e) {
  return {
    get value() {
      {
        const r = e();
        return Object.defineProperty(this, "value", { value: r }), r;
      }
    }
  };
}
function Hn(e) {
  return e == null;
}
function Gn(e) {
  const r = e.startsWith("^") ? 1 : 0, n = e.endsWith("$") ? e.length - 1 : e.length;
  return e.slice(r, n);
}
function Xi(e, r) {
  const n = (e.toString().split(".")[1] || "").length, t = r.toString();
  let o = (t.split(".")[1] || "").length;
  if (o === 0 && /\d?e-\d?/.test(t)) {
    const l = t.match(/\d?e-(\d?)/);
    l?.[1] && (o = Number.parseInt(l[1]));
  }
  const u = n > o ? n : o, c = Number.parseInt(e.toFixed(u).replace(".", "")), a = Number.parseInt(r.toFixed(u).replace(".", ""));
  return c % a / 10 ** u;
}
const Nt = Symbol("evaluating");
function k(e, r, n) {
  let t;
  Object.defineProperty(e, r, {
    get() {
      if (t !== Nt)
        return t === void 0 && (t = Nt, t = n()), t;
    },
    set(o) {
      Object.defineProperty(e, r, {
        value: o
        // configurable: true,
      });
    },
    configurable: !0
  });
}
function Me(e, r, n) {
  Object.defineProperty(e, r, {
    value: n,
    writable: !0,
    enumerable: !0,
    configurable: !0
  });
}
function We(...e) {
  const r = {};
  for (const n of e) {
    const t = Object.getOwnPropertyDescriptors(n);
    Object.assign(r, t);
  }
  return Object.defineProperties({}, r);
}
function Ut(e) {
  return JSON.stringify(e);
}
const vo = "captureStackTrace" in Error ? Error.captureStackTrace : (...e) => {
};
function Yr(e) {
  return typeof e == "object" && e !== null && !Array.isArray(e);
}
const Qi = Vn(() => {
  if (typeof navigator < "u" && navigator?.userAgent?.includes("Cloudflare"))
    return !1;
  try {
    const e = Function;
    return new e(""), !0;
  } catch {
    return !1;
  }
});
function wr(e) {
  if (Yr(e) === !1)
    return !1;
  const r = e.constructor;
  if (r === void 0)
    return !0;
  const n = r.prototype;
  return !(Yr(n) === !1 || Object.prototype.hasOwnProperty.call(n, "isPrototypeOf") === !1);
}
function Oo(e) {
  return wr(e) ? { ...e } : Array.isArray(e) ? [...e] : e;
}
const eu = /* @__PURE__ */ new Set(["string", "number", "symbol"]);
function tr(e) {
  return e.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}
function xe(e, r, n) {
  const t = new e._zod.constr(r ?? e._zod.def);
  return (!r || n?.parent) && (t._zod.parent = e), t;
}
function h(e) {
  const r = e;
  if (!r)
    return {};
  if (typeof r == "string")
    return { error: () => r };
  if (r?.message !== void 0) {
    if (r?.error !== void 0)
      throw new Error("Cannot specify both `message` and `error` params");
    r.error = r.message;
  }
  return delete r.message, typeof r.error == "string" ? { ...r, error: () => r.error } : r;
}
function ru(e) {
  return Object.keys(e).filter((r) => e[r]._zod.optin === "optional" && e[r]._zod.optout === "optional");
}
const nu = {
  safeint: [Number.MIN_SAFE_INTEGER, Number.MAX_SAFE_INTEGER],
  int32: [-2147483648, 2147483647],
  uint32: [0, 4294967295],
  float32: [-34028234663852886e22, 34028234663852886e22],
  float64: [-Number.MAX_VALUE, Number.MAX_VALUE]
};
function tu(e, r) {
  const n = e._zod.def, t = We(e._zod.def, {
    get shape() {
      const o = {};
      for (const u in r) {
        if (!(u in n.shape))
          throw new Error(`Unrecognized key: "${u}"`);
        r[u] && (o[u] = n.shape[u]);
      }
      return Me(this, "shape", o), o;
    },
    checks: []
  });
  return xe(e, t);
}
function ou(e, r) {
  const n = e._zod.def, t = We(e._zod.def, {
    get shape() {
      const o = { ...e._zod.def.shape };
      for (const u in r) {
        if (!(u in n.shape))
          throw new Error(`Unrecognized key: "${u}"`);
        r[u] && delete o[u];
      }
      return Me(this, "shape", o), o;
    },
    checks: []
  });
  return xe(e, t);
}
function iu(e, r) {
  if (!wr(r))
    throw new Error("Invalid input to extend: expected a plain object");
  const n = e._zod.def.checks;
  if (n && n.length > 0)
    throw new Error("Object schemas containing refinements cannot be extended. Use `.safeExtend()` instead.");
  const o = We(e._zod.def, {
    get shape() {
      const u = { ...e._zod.def.shape, ...r };
      return Me(this, "shape", u), u;
    },
    checks: []
  });
  return xe(e, o);
}
function uu(e, r) {
  if (!wr(r))
    throw new Error("Invalid input to safeExtend: expected a plain object");
  const n = {
    ...e._zod.def,
    get shape() {
      const t = { ...e._zod.def.shape, ...r };
      return Me(this, "shape", t), t;
    },
    checks: e._zod.def.checks
  };
  return xe(e, n);
}
function cu(e, r) {
  const n = We(e._zod.def, {
    get shape() {
      const t = { ...e._zod.def.shape, ...r._zod.def.shape };
      return Me(this, "shape", t), t;
    },
    get catchall() {
      return r._zod.def.catchall;
    },
    checks: []
    // delete existing checks
  });
  return xe(e, n);
}
function au(e, r, n) {
  const t = We(r._zod.def, {
    get shape() {
      const o = r._zod.def.shape, u = { ...o };
      if (n)
        for (const c in n) {
          if (!(c in o))
            throw new Error(`Unrecognized key: "${c}"`);
          n[c] && (u[c] = e ? new e({
            type: "optional",
            innerType: o[c]
          }) : o[c]);
        }
      else
        for (const c in o)
          u[c] = e ? new e({
            type: "optional",
            innerType: o[c]
          }) : o[c];
      return Me(this, "shape", u), u;
    },
    checks: []
  });
  return xe(r, t);
}
function su(e, r, n) {
  const t = We(r._zod.def, {
    get shape() {
      const o = r._zod.def.shape, u = { ...o };
      if (n)
        for (const c in n) {
          if (!(c in u))
            throw new Error(`Unrecognized key: "${c}"`);
          n[c] && (u[c] = new e({
            type: "nonoptional",
            innerType: o[c]
          }));
        }
      else
        for (const c in o)
          u[c] = new e({
            type: "nonoptional",
            innerType: o[c]
          });
      return Me(this, "shape", u), u;
    },
    checks: []
  });
  return xe(r, t);
}
function er(e, r = 0) {
  if (e.aborted === !0)
    return !0;
  for (let n = r; n < e.issues.length; n++)
    if (e.issues[n]?.continue !== !0)
      return !0;
  return !1;
}
function $o(e, r) {
  return r.map((n) => {
    var t;
    return (t = n).path ?? (t.path = []), n.path.unshift(e), n;
  });
}
function Kr(e) {
  return typeof e == "string" ? e : e?.message;
}
function Ue(e, r, n) {
  const t = { ...e, path: e.path ?? [] };
  if (!e.message) {
    const o = Kr(e.inst?._zod.def?.error?.(e)) ?? Kr(r?.error?.(e)) ?? Kr(n.customError?.(e)) ?? Kr(n.localeError?.(e)) ?? "Invalid input";
    t.message = o;
  }
  return delete t.inst, delete t.continue, r?.reportInput || delete t.input, t;
}
function qn(e) {
  return Array.isArray(e) ? "array" : typeof e == "string" ? "string" : "unknown";
}
function yr(...e) {
  const [r, n, t] = e;
  return typeof r == "string" ? {
    message: r,
    code: "custom",
    input: n,
    inst: t
  } : { ...r };
}
const _o = (e, r) => {
  e.name = "$ZodError", Object.defineProperty(e, "_zod", {
    value: e._zod,
    enumerable: !1
  }), Object.defineProperty(e, "issues", {
    value: r,
    enumerable: !1
  }), e.message = JSON.stringify(r, En, 2), Object.defineProperty(e, "toString", {
    value: () => e.message,
    enumerable: !1
  });
}, wo = f("$ZodError", _o), yo = f("$ZodError", _o, { Parent: Error });
function fu(e, r = (n) => n.message) {
  const n = {}, t = [];
  for (const o of e.issues)
    o.path.length > 0 ? (n[o.path[0]] = n[o.path[0]] || [], n[o.path[0]].push(r(o))) : t.push(r(o));
  return { formErrors: t, fieldErrors: n };
}
function du(e, r = (n) => n.message) {
  const n = { _errors: [] }, t = (o) => {
    for (const u of o.issues)
      if (u.code === "invalid_union" && u.errors.length)
        u.errors.map((c) => t({ issues: c }));
      else if (u.code === "invalid_key")
        t({ issues: u.issues });
      else if (u.code === "invalid_element")
        t({ issues: u.issues });
      else if (u.path.length === 0)
        n._errors.push(r(u));
      else {
        let c = n, a = 0;
        for (; a < u.path.length; ) {
          const l = u.path[a];
          a === u.path.length - 1 ? (c[l] = c[l] || { _errors: [] }, c[l]._errors.push(r(u))) : c[l] = c[l] || { _errors: [] }, c = c[l], a++;
        }
      }
  };
  return t(e), n;
}
const Jn = (e) => (r, n, t, o) => {
  const u = t ? Object.assign(t, { async: !1 }) : { async: !1 }, c = r._zod.run({ value: n, issues: [] }, u);
  if (c instanceof Promise)
    throw new rr();
  if (c.issues.length) {
    const a = new (o?.Err ?? e)(c.issues.map((l) => Ue(l, u, Ne())));
    throw vo(a, o?.callee), a;
  }
  return c.value;
}, Yn = (e) => async (r, n, t, o) => {
  const u = t ? Object.assign(t, { async: !0 }) : { async: !0 };
  let c = r._zod.run({ value: n, issues: [] }, u);
  if (c instanceof Promise && (c = await c), c.issues.length) {
    const a = new (o?.Err ?? e)(c.issues.map((l) => Ue(l, u, Ne())));
    throw vo(a, o?.callee), a;
  }
  return c.value;
}, on = (e) => (r, n, t) => {
  const o = t ? { ...t, async: !1 } : { async: !1 }, u = r._zod.run({ value: n, issues: [] }, o);
  if (u instanceof Promise)
    throw new rr();
  return u.issues.length ? {
    success: !1,
    error: new (e ?? wo)(u.issues.map((c) => Ue(c, o, Ne())))
  } : { success: !0, data: u.value };
}, mu = /* @__PURE__ */ on(yo), un = (e) => async (r, n, t) => {
  const o = t ? Object.assign(t, { async: !0 }) : { async: !0 };
  let u = r._zod.run({ value: n, issues: [] }, o);
  return u instanceof Promise && (u = await u), u.issues.length ? {
    success: !1,
    error: new e(u.issues.map((c) => Ue(c, o, Ne())))
  } : { success: !0, data: u.value };
}, lu = /* @__PURE__ */ un(yo), pu = (e) => (r, n, t) => {
  const o = t ? Object.assign(t, { direction: "backward" }) : { direction: "backward" };
  return Jn(e)(r, n, o);
}, hu = (e) => (r, n, t) => Jn(e)(r, n, t), gu = (e) => async (r, n, t) => {
  const o = t ? Object.assign(t, { direction: "backward" }) : { direction: "backward" };
  return Yn(e)(r, n, o);
}, bu = (e) => async (r, n, t) => Yn(e)(r, n, t), Fu = (e) => (r, n, t) => {
  const o = t ? Object.assign(t, { direction: "backward" }) : { direction: "backward" };
  return on(e)(r, n, o);
}, vu = (e) => (r, n, t) => on(e)(r, n, t), Ou = (e) => async (r, n, t) => {
  const o = t ? Object.assign(t, { direction: "backward" }) : { direction: "backward" };
  return un(e)(r, n, o);
}, $u = (e) => async (r, n, t) => un(e)(r, n, t), _u = /^[cC][^\s-]{8,}$/, wu = /^[0-9a-z]+$/, yu = /^[0-9A-HJKMNP-TV-Za-hjkmnp-tv-z]{26}$/, ku = /^[0-9a-vA-V]{20}$/, Iu = /^[A-Za-z0-9]{27}$/, Ru = /^[a-zA-Z0-9_-]{21}$/, xu = /^P(?:(\d+W)|(?!.*W)(?=\d|T\d)(\d+Y)?(\d+M)?(\d+D)?(T(?=\d)(\d+H)?(\d+M)?(\d+([.,]\d+)?S)?)?)$/, Au = /^([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12})$/, Dt = (e) => e ? new RegExp(`^([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-${e}[0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12})$`) : /^([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[1-8][0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}|00000000-0000-0000-0000-000000000000|ffffffff-ffff-ffff-ffff-ffffffffffff)$/, zu = /^(?!\.)(?!.*\.\.)([A-Za-z0-9_'+\-\.]*)[A-Za-z0-9_+-]@([A-Za-z0-9][A-Za-z0-9\-]*\.)+[A-Za-z]{2,}$/, Tu = "^(\\p{Extended_Pictographic}|\\p{Emoji_Component})+$";
function Pu() {
  return new RegExp(Tu, "u");
}
const Su = /^(?:(?:25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9][0-9]|[0-9])\.){3}(?:25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9][0-9]|[0-9])$/, Zu = /^(([0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:))$/, ju = /^((25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9][0-9]|[0-9])\.){3}(25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9][0-9]|[0-9])\/([0-9]|[1-2][0-9]|3[0-2])$/, Eu = /^(([0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}|::|([0-9a-fA-F]{1,4})?::([0-9a-fA-F]{1,4}:?){0,6})\/(12[0-8]|1[01][0-9]|[1-9]?[0-9])$/, Cu = /^$|^(?:[0-9a-zA-Z+/]{4})*(?:(?:[0-9a-zA-Z+/]{2}==)|(?:[0-9a-zA-Z+/]{3}=))?$/, ko = /^[A-Za-z0-9_-]*$/, Nu = /^(?=.{1,253}\.?$)[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[-0-9a-zA-Z]{0,61}[0-9a-zA-Z])?)*\.?$/, Uu = /^\+(?:[0-9]){6,14}[0-9]$/, Io = "(?:(?:\\d\\d[2468][048]|\\d\\d[13579][26]|\\d\\d0[48]|[02468][048]00|[13579][26]00)-02-29|\\d{4}-(?:(?:0[13578]|1[02])-(?:0[1-9]|[12]\\d|3[01])|(?:0[469]|11)-(?:0[1-9]|[12]\\d|30)|(?:02)-(?:0[1-9]|1\\d|2[0-8])))", Du = /* @__PURE__ */ new RegExp(`^${Io}$`);
function Ro(e) {
  const r = "(?:[01]\\d|2[0-3]):[0-5]\\d";
  return typeof e.precision == "number" ? e.precision === -1 ? `${r}` : e.precision === 0 ? `${r}:[0-5]\\d` : `${r}:[0-5]\\d\\.\\d{${e.precision}}` : `${r}(?::[0-5]\\d(?:\\.\\d+)?)?`;
}
function Lu(e) {
  return new RegExp(`^${Ro(e)}$`);
}
function Mu(e) {
  const r = Ro({ precision: e.precision }), n = ["Z"];
  e.local && n.push(""), e.offset && n.push("([+-](?:[01]\\d|2[0-3]):[0-5]\\d)");
  const t = `${r}(?:${n.join("|")})`;
  return new RegExp(`^${Io}T(?:${t})$`);
}
const Wu = (e) => {
  const r = e ? `[\\s\\S]{${e?.minimum ?? 0},${e?.maximum ?? ""}}` : "[\\s\\S]*";
  return new RegExp(`^${r}$`);
}, Bu = /^-?\d+$/, Ku = /^-?\d+(?:\.\d+)?/, Vu = /^(?:true|false)$/i, Hu = /^[^A-Z]*$/, Gu = /^[^a-z]*$/, Y = /* @__PURE__ */ f("$ZodCheck", (e, r) => {
  var n;
  e._zod ?? (e._zod = {}), e._zod.def = r, (n = e._zod).onattach ?? (n.onattach = []);
}), xo = {
  number: "number",
  bigint: "bigint",
  object: "date"
}, Ao = /* @__PURE__ */ f("$ZodCheckLessThan", (e, r) => {
  Y.init(e, r);
  const n = xo[typeof r.value];
  e._zod.onattach.push((t) => {
    const o = t._zod.bag, u = (r.inclusive ? o.maximum : o.exclusiveMaximum) ?? Number.POSITIVE_INFINITY;
    r.value < u && (r.inclusive ? o.maximum = r.value : o.exclusiveMaximum = r.value);
  }), e._zod.check = (t) => {
    (r.inclusive ? t.value <= r.value : t.value < r.value) || t.issues.push({
      origin: n,
      code: "too_big",
      maximum: r.value,
      input: t.value,
      inclusive: r.inclusive,
      inst: e,
      continue: !r.abort
    });
  };
}), zo = /* @__PURE__ */ f("$ZodCheckGreaterThan", (e, r) => {
  Y.init(e, r);
  const n = xo[typeof r.value];
  e._zod.onattach.push((t) => {
    const o = t._zod.bag, u = (r.inclusive ? o.minimum : o.exclusiveMinimum) ?? Number.NEGATIVE_INFINITY;
    r.value > u && (r.inclusive ? o.minimum = r.value : o.exclusiveMinimum = r.value);
  }), e._zod.check = (t) => {
    (r.inclusive ? t.value >= r.value : t.value > r.value) || t.issues.push({
      origin: n,
      code: "too_small",
      minimum: r.value,
      input: t.value,
      inclusive: r.inclusive,
      inst: e,
      continue: !r.abort
    });
  };
}), qu = /* @__PURE__ */ f("$ZodCheckMultipleOf", (e, r) => {
  Y.init(e, r), e._zod.onattach.push((n) => {
    var t;
    (t = n._zod.bag).multipleOf ?? (t.multipleOf = r.value);
  }), e._zod.check = (n) => {
    if (typeof n.value != typeof r.value)
      throw new Error("Cannot mix number and bigint in multiple_of check.");
    (typeof n.value == "bigint" ? n.value % r.value === BigInt(0) : Xi(n.value, r.value) === 0) || n.issues.push({
      origin: typeof n.value,
      code: "not_multiple_of",
      divisor: r.value,
      input: n.value,
      inst: e,
      continue: !r.abort
    });
  };
}), Ju = /* @__PURE__ */ f("$ZodCheckNumberFormat", (e, r) => {
  Y.init(e, r), r.format = r.format || "float64";
  const n = r.format?.includes("int"), t = n ? "int" : "number", [o, u] = nu[r.format];
  e._zod.onattach.push((c) => {
    const a = c._zod.bag;
    a.format = r.format, a.minimum = o, a.maximum = u, n && (a.pattern = Bu);
  }), e._zod.check = (c) => {
    const a = c.value;
    if (n) {
      if (!Number.isInteger(a)) {
        c.issues.push({
          expected: t,
          format: r.format,
          code: "invalid_type",
          continue: !1,
          input: a,
          inst: e
        });
        return;
      }
      if (!Number.isSafeInteger(a)) {
        a > 0 ? c.issues.push({
          input: a,
          code: "too_big",
          maximum: Number.MAX_SAFE_INTEGER,
          note: "Integers must be within the safe integer range.",
          inst: e,
          origin: t,
          continue: !r.abort
        }) : c.issues.push({
          input: a,
          code: "too_small",
          minimum: Number.MIN_SAFE_INTEGER,
          note: "Integers must be within the safe integer range.",
          inst: e,
          origin: t,
          continue: !r.abort
        });
        return;
      }
    }
    a < o && c.issues.push({
      origin: "number",
      input: a,
      code: "too_small",
      minimum: o,
      inclusive: !0,
      inst: e,
      continue: !r.abort
    }), a > u && c.issues.push({
      origin: "number",
      input: a,
      code: "too_big",
      maximum: u,
      inst: e
    });
  };
}), Yu = /* @__PURE__ */ f("$ZodCheckMaxLength", (e, r) => {
  var n;
  Y.init(e, r), (n = e._zod.def).when ?? (n.when = (t) => {
    const o = t.value;
    return !Hn(o) && o.length !== void 0;
  }), e._zod.onattach.push((t) => {
    const o = t._zod.bag.maximum ?? Number.POSITIVE_INFINITY;
    r.maximum < o && (t._zod.bag.maximum = r.maximum);
  }), e._zod.check = (t) => {
    const o = t.value;
    if (o.length <= r.maximum)
      return;
    const c = qn(o);
    t.issues.push({
      origin: c,
      code: "too_big",
      maximum: r.maximum,
      inclusive: !0,
      input: o,
      inst: e,
      continue: !r.abort
    });
  };
}), Xu = /* @__PURE__ */ f("$ZodCheckMinLength", (e, r) => {
  var n;
  Y.init(e, r), (n = e._zod.def).when ?? (n.when = (t) => {
    const o = t.value;
    return !Hn(o) && o.length !== void 0;
  }), e._zod.onattach.push((t) => {
    const o = t._zod.bag.minimum ?? Number.NEGATIVE_INFINITY;
    r.minimum > o && (t._zod.bag.minimum = r.minimum);
  }), e._zod.check = (t) => {
    const o = t.value;
    if (o.length >= r.minimum)
      return;
    const c = qn(o);
    t.issues.push({
      origin: c,
      code: "too_small",
      minimum: r.minimum,
      inclusive: !0,
      input: o,
      inst: e,
      continue: !r.abort
    });
  };
}), Qu = /* @__PURE__ */ f("$ZodCheckLengthEquals", (e, r) => {
  var n;
  Y.init(e, r), (n = e._zod.def).when ?? (n.when = (t) => {
    const o = t.value;
    return !Hn(o) && o.length !== void 0;
  }), e._zod.onattach.push((t) => {
    const o = t._zod.bag;
    o.minimum = r.length, o.maximum = r.length, o.length = r.length;
  }), e._zod.check = (t) => {
    const o = t.value, u = o.length;
    if (u === r.length)
      return;
    const c = qn(o), a = u > r.length;
    t.issues.push({
      origin: c,
      ...a ? { code: "too_big", maximum: r.length } : { code: "too_small", minimum: r.length },
      inclusive: !0,
      exact: !0,
      input: t.value,
      inst: e,
      continue: !r.abort
    });
  };
}), cn = /* @__PURE__ */ f("$ZodCheckStringFormat", (e, r) => {
  var n, t;
  Y.init(e, r), e._zod.onattach.push((o) => {
    const u = o._zod.bag;
    u.format = r.format, r.pattern && (u.patterns ?? (u.patterns = /* @__PURE__ */ new Set()), u.patterns.add(r.pattern));
  }), r.pattern ? (n = e._zod).check ?? (n.check = (o) => {
    r.pattern.lastIndex = 0, !r.pattern.test(o.value) && o.issues.push({
      origin: "string",
      code: "invalid_format",
      format: r.format,
      input: o.value,
      ...r.pattern ? { pattern: r.pattern.toString() } : {},
      inst: e,
      continue: !r.abort
    });
  }) : (t = e._zod).check ?? (t.check = () => {
  });
}), ec = /* @__PURE__ */ f("$ZodCheckRegex", (e, r) => {
  cn.init(e, r), e._zod.check = (n) => {
    r.pattern.lastIndex = 0, !r.pattern.test(n.value) && n.issues.push({
      origin: "string",
      code: "invalid_format",
      format: "regex",
      input: n.value,
      pattern: r.pattern.toString(),
      inst: e,
      continue: !r.abort
    });
  };
}), rc = /* @__PURE__ */ f("$ZodCheckLowerCase", (e, r) => {
  r.pattern ?? (r.pattern = Hu), cn.init(e, r);
}), nc = /* @__PURE__ */ f("$ZodCheckUpperCase", (e, r) => {
  r.pattern ?? (r.pattern = Gu), cn.init(e, r);
}), tc = /* @__PURE__ */ f("$ZodCheckIncludes", (e, r) => {
  Y.init(e, r);
  const n = tr(r.includes), t = new RegExp(typeof r.position == "number" ? `^.{${r.position}}${n}` : n);
  r.pattern = t, e._zod.onattach.push((o) => {
    const u = o._zod.bag;
    u.patterns ?? (u.patterns = /* @__PURE__ */ new Set()), u.patterns.add(t);
  }), e._zod.check = (o) => {
    o.value.includes(r.includes, r.position) || o.issues.push({
      origin: "string",
      code: "invalid_format",
      format: "includes",
      includes: r.includes,
      input: o.value,
      inst: e,
      continue: !r.abort
    });
  };
}), oc = /* @__PURE__ */ f("$ZodCheckStartsWith", (e, r) => {
  Y.init(e, r);
  const n = new RegExp(`^${tr(r.prefix)}.*`);
  r.pattern ?? (r.pattern = n), e._zod.onattach.push((t) => {
    const o = t._zod.bag;
    o.patterns ?? (o.patterns = /* @__PURE__ */ new Set()), o.patterns.add(n);
  }), e._zod.check = (t) => {
    t.value.startsWith(r.prefix) || t.issues.push({
      origin: "string",
      code: "invalid_format",
      format: "starts_with",
      prefix: r.prefix,
      input: t.value,
      inst: e,
      continue: !r.abort
    });
  };
}), ic = /* @__PURE__ */ f("$ZodCheckEndsWith", (e, r) => {
  Y.init(e, r);
  const n = new RegExp(`.*${tr(r.suffix)}$`);
  r.pattern ?? (r.pattern = n), e._zod.onattach.push((t) => {
    const o = t._zod.bag;
    o.patterns ?? (o.patterns = /* @__PURE__ */ new Set()), o.patterns.add(n);
  }), e._zod.check = (t) => {
    t.value.endsWith(r.suffix) || t.issues.push({
      origin: "string",
      code: "invalid_format",
      format: "ends_with",
      suffix: r.suffix,
      input: t.value,
      inst: e,
      continue: !r.abort
    });
  };
}), uc = /* @__PURE__ */ f("$ZodCheckOverwrite", (e, r) => {
  Y.init(e, r), e._zod.check = (n) => {
    n.value = r.tx(n.value);
  };
});
class cc {
  constructor(r = []) {
    this.content = [], this.indent = 0, this && (this.args = r);
  }
  indented(r) {
    this.indent += 1, r(this), this.indent -= 1;
  }
  write(r) {
    if (typeof r == "function") {
      r(this, { execution: "sync" }), r(this, { execution: "async" });
      return;
    }
    const t = r.split(`
`).filter((c) => c), o = Math.min(...t.map((c) => c.length - c.trimStart().length)), u = t.map((c) => c.slice(o)).map((c) => " ".repeat(this.indent * 2) + c);
    for (const c of u)
      this.content.push(c);
  }
  compile() {
    const r = Function, n = this?.args, o = [...(this?.content ?? [""]).map((u) => `  ${u}`)];
    return new r(...n, o.join(`
`));
  }
}
const ac = {
  major: 4,
  minor: 1,
  patch: 12
}, z = /* @__PURE__ */ f("$ZodType", (e, r) => {
  var n;
  e ?? (e = {}), e._zod.def = r, e._zod.bag = e._zod.bag || {}, e._zod.version = ac;
  const t = [...e._zod.def.checks ?? []];
  e._zod.traits.has("$ZodCheck") && t.unshift(e);
  for (const o of t)
    for (const u of o._zod.onattach)
      u(e);
  if (t.length === 0)
    (n = e._zod).deferred ?? (n.deferred = []), e._zod.deferred?.push(() => {
      e._zod.run = e._zod.parse;
    });
  else {
    const o = (c, a, l) => {
      let p = er(c), b;
      for (const y of a) {
        if (y._zod.def.when) {
          if (!y._zod.def.when(c))
            continue;
        } else if (p)
          continue;
        const O = c.issues.length, I = y._zod.check(c);
        if (I instanceof Promise && l?.async === !1)
          throw new rr();
        if (b || I instanceof Promise)
          b = (b ?? Promise.resolve()).then(async () => {
            await I, c.issues.length !== O && (p || (p = er(c, O)));
          });
        else {
          if (c.issues.length === O)
            continue;
          p || (p = er(c, O));
        }
      }
      return b ? b.then(() => c) : c;
    }, u = (c, a, l) => {
      if (er(c))
        return c.aborted = !0, c;
      const p = o(a, t, l);
      if (p instanceof Promise) {
        if (l.async === !1)
          throw new rr();
        return p.then((b) => e._zod.parse(b, l));
      }
      return e._zod.parse(p, l);
    };
    e._zod.run = (c, a) => {
      if (a.skipChecks)
        return e._zod.parse(c, a);
      if (a.direction === "backward") {
        const p = e._zod.parse({ value: c.value, issues: [] }, { ...a, skipChecks: !0 });
        return p instanceof Promise ? p.then((b) => u(b, c, a)) : u(p, c, a);
      }
      const l = e._zod.parse(c, a);
      if (l instanceof Promise) {
        if (a.async === !1)
          throw new rr();
        return l.then((p) => o(p, t, a));
      }
      return o(l, t, a);
    };
  }
  e["~standard"] = {
    validate: (o) => {
      try {
        const u = mu(e, o);
        return u.success ? { value: u.data } : { issues: u.error?.issues };
      } catch {
        return lu(e, o).then((c) => c.success ? { value: c.data } : { issues: c.error?.issues });
      }
    },
    vendor: "zod",
    version: 1
  };
}), Xn = /* @__PURE__ */ f("$ZodString", (e, r) => {
  z.init(e, r), e._zod.pattern = [...e?._zod.bag?.patterns ?? []].pop() ?? Wu(e._zod.bag), e._zod.parse = (n, t) => {
    if (r.coerce)
      try {
        n.value = String(n.value);
      } catch {
      }
    return typeof n.value == "string" || n.issues.push({
      expected: "string",
      code: "invalid_type",
      input: n.value,
      inst: e
    }), n;
  };
}), R = /* @__PURE__ */ f("$ZodStringFormat", (e, r) => {
  cn.init(e, r), Xn.init(e, r);
}), sc = /* @__PURE__ */ f("$ZodGUID", (e, r) => {
  r.pattern ?? (r.pattern = Au), R.init(e, r);
}), fc = /* @__PURE__ */ f("$ZodUUID", (e, r) => {
  if (r.version) {
    const t = {
      v1: 1,
      v2: 2,
      v3: 3,
      v4: 4,
      v5: 5,
      v6: 6,
      v7: 7,
      v8: 8
    }[r.version];
    if (t === void 0)
      throw new Error(`Invalid UUID version: "${r.version}"`);
    r.pattern ?? (r.pattern = Dt(t));
  } else
    r.pattern ?? (r.pattern = Dt());
  R.init(e, r);
}), dc = /* @__PURE__ */ f("$ZodEmail", (e, r) => {
  r.pattern ?? (r.pattern = zu), R.init(e, r);
}), mc = /* @__PURE__ */ f("$ZodURL", (e, r) => {
  R.init(e, r), e._zod.check = (n) => {
    try {
      const t = n.value.trim(), o = new URL(t);
      r.hostname && (r.hostname.lastIndex = 0, r.hostname.test(o.hostname) || n.issues.push({
        code: "invalid_format",
        format: "url",
        note: "Invalid hostname",
        pattern: Nu.source,
        input: n.value,
        inst: e,
        continue: !r.abort
      })), r.protocol && (r.protocol.lastIndex = 0, r.protocol.test(o.protocol.endsWith(":") ? o.protocol.slice(0, -1) : o.protocol) || n.issues.push({
        code: "invalid_format",
        format: "url",
        note: "Invalid protocol",
        pattern: r.protocol.source,
        input: n.value,
        inst: e,
        continue: !r.abort
      })), r.normalize ? n.value = o.href : n.value = t;
      return;
    } catch {
      n.issues.push({
        code: "invalid_format",
        format: "url",
        input: n.value,
        inst: e,
        continue: !r.abort
      });
    }
  };
}), lc = /* @__PURE__ */ f("$ZodEmoji", (e, r) => {
  r.pattern ?? (r.pattern = Pu()), R.init(e, r);
}), pc = /* @__PURE__ */ f("$ZodNanoID", (e, r) => {
  r.pattern ?? (r.pattern = Ru), R.init(e, r);
}), hc = /* @__PURE__ */ f("$ZodCUID", (e, r) => {
  r.pattern ?? (r.pattern = _u), R.init(e, r);
}), gc = /* @__PURE__ */ f("$ZodCUID2", (e, r) => {
  r.pattern ?? (r.pattern = wu), R.init(e, r);
}), bc = /* @__PURE__ */ f("$ZodULID", (e, r) => {
  r.pattern ?? (r.pattern = yu), R.init(e, r);
}), Fc = /* @__PURE__ */ f("$ZodXID", (e, r) => {
  r.pattern ?? (r.pattern = ku), R.init(e, r);
}), vc = /* @__PURE__ */ f("$ZodKSUID", (e, r) => {
  r.pattern ?? (r.pattern = Iu), R.init(e, r);
}), Oc = /* @__PURE__ */ f("$ZodISODateTime", (e, r) => {
  r.pattern ?? (r.pattern = Mu(r)), R.init(e, r);
}), $c = /* @__PURE__ */ f("$ZodISODate", (e, r) => {
  r.pattern ?? (r.pattern = Du), R.init(e, r);
}), _c = /* @__PURE__ */ f("$ZodISOTime", (e, r) => {
  r.pattern ?? (r.pattern = Lu(r)), R.init(e, r);
}), wc = /* @__PURE__ */ f("$ZodISODuration", (e, r) => {
  r.pattern ?? (r.pattern = xu), R.init(e, r);
}), yc = /* @__PURE__ */ f("$ZodIPv4", (e, r) => {
  r.pattern ?? (r.pattern = Su), R.init(e, r), e._zod.onattach.push((n) => {
    const t = n._zod.bag;
    t.format = "ipv4";
  });
}), kc = /* @__PURE__ */ f("$ZodIPv6", (e, r) => {
  r.pattern ?? (r.pattern = Zu), R.init(e, r), e._zod.onattach.push((n) => {
    const t = n._zod.bag;
    t.format = "ipv6";
  }), e._zod.check = (n) => {
    try {
      new URL(`http://[${n.value}]`);
    } catch {
      n.issues.push({
        code: "invalid_format",
        format: "ipv6",
        input: n.value,
        inst: e,
        continue: !r.abort
      });
    }
  };
}), Ic = /* @__PURE__ */ f("$ZodCIDRv4", (e, r) => {
  r.pattern ?? (r.pattern = ju), R.init(e, r);
}), Rc = /* @__PURE__ */ f("$ZodCIDRv6", (e, r) => {
  r.pattern ?? (r.pattern = Eu), R.init(e, r), e._zod.check = (n) => {
    const t = n.value.split("/");
    try {
      if (t.length !== 2)
        throw new Error();
      const [o, u] = t;
      if (!u)
        throw new Error();
      const c = Number(u);
      if (`${c}` !== u)
        throw new Error();
      if (c < 0 || c > 128)
        throw new Error();
      new URL(`http://[${o}]`);
    } catch {
      n.issues.push({
        code: "invalid_format",
        format: "cidrv6",
        input: n.value,
        inst: e,
        continue: !r.abort
      });
    }
  };
});
function To(e) {
  if (e === "")
    return !0;
  if (e.length % 4 !== 0)
    return !1;
  try {
    return atob(e), !0;
  } catch {
    return !1;
  }
}
const xc = /* @__PURE__ */ f("$ZodBase64", (e, r) => {
  r.pattern ?? (r.pattern = Cu), R.init(e, r), e._zod.onattach.push((n) => {
    n._zod.bag.contentEncoding = "base64";
  }), e._zod.check = (n) => {
    To(n.value) || n.issues.push({
      code: "invalid_format",
      format: "base64",
      input: n.value,
      inst: e,
      continue: !r.abort
    });
  };
});
function Ac(e) {
  if (!ko.test(e))
    return !1;
  const r = e.replace(/[-_]/g, (t) => t === "-" ? "+" : "/"), n = r.padEnd(Math.ceil(r.length / 4) * 4, "=");
  return To(n);
}
const zc = /* @__PURE__ */ f("$ZodBase64URL", (e, r) => {
  r.pattern ?? (r.pattern = ko), R.init(e, r), e._zod.onattach.push((n) => {
    n._zod.bag.contentEncoding = "base64url";
  }), e._zod.check = (n) => {
    Ac(n.value) || n.issues.push({
      code: "invalid_format",
      format: "base64url",
      input: n.value,
      inst: e,
      continue: !r.abort
    });
  };
}), Tc = /* @__PURE__ */ f("$ZodE164", (e, r) => {
  r.pattern ?? (r.pattern = Uu), R.init(e, r);
});
function Pc(e, r = null) {
  try {
    const n = e.split(".");
    if (n.length !== 3)
      return !1;
    const [t] = n;
    if (!t)
      return !1;
    const o = JSON.parse(atob(t));
    return !("typ" in o && o?.typ !== "JWT" || !o.alg || r && (!("alg" in o) || o.alg !== r));
  } catch {
    return !1;
  }
}
const Sc = /* @__PURE__ */ f("$ZodJWT", (e, r) => {
  R.init(e, r), e._zod.check = (n) => {
    Pc(n.value, r.alg) || n.issues.push({
      code: "invalid_format",
      format: "jwt",
      input: n.value,
      inst: e,
      continue: !r.abort
    });
  };
}), Po = /* @__PURE__ */ f("$ZodNumber", (e, r) => {
  z.init(e, r), e._zod.pattern = e._zod.bag.pattern ?? Ku, e._zod.parse = (n, t) => {
    if (r.coerce)
      try {
        n.value = Number(n.value);
      } catch {
      }
    const o = n.value;
    if (typeof o == "number" && !Number.isNaN(o) && Number.isFinite(o))
      return n;
    const u = typeof o == "number" ? Number.isNaN(o) ? "NaN" : Number.isFinite(o) ? void 0 : "Infinity" : void 0;
    return n.issues.push({
      expected: "number",
      code: "invalid_type",
      input: o,
      inst: e,
      ...u ? { received: u } : {}
    }), n;
  };
}), Zc = /* @__PURE__ */ f("$ZodNumber", (e, r) => {
  Ju.init(e, r), Po.init(e, r);
}), jc = /* @__PURE__ */ f("$ZodBoolean", (e, r) => {
  z.init(e, r), e._zod.pattern = Vu, e._zod.parse = (n, t) => {
    if (r.coerce)
      try {
        n.value = !!n.value;
      } catch {
      }
    const o = n.value;
    return typeof o == "boolean" || n.issues.push({
      expected: "boolean",
      code: "invalid_type",
      input: o,
      inst: e
    }), n;
  };
}), Ec = /* @__PURE__ */ f("$ZodUnknown", (e, r) => {
  z.init(e, r), e._zod.parse = (n) => n;
}), Cc = /* @__PURE__ */ f("$ZodNever", (e, r) => {
  z.init(e, r), e._zod.parse = (n, t) => (n.issues.push({
    expected: "never",
    code: "invalid_type",
    input: n.value,
    inst: e
  }), n);
});
function Lt(e, r, n) {
  e.issues.length && r.issues.push(...$o(n, e.issues)), r.value[n] = e.value;
}
const Nc = /* @__PURE__ */ f("$ZodArray", (e, r) => {
  z.init(e, r), e._zod.parse = (n, t) => {
    const o = n.value;
    if (!Array.isArray(o))
      return n.issues.push({
        expected: "array",
        code: "invalid_type",
        input: o,
        inst: e
      }), n;
    n.value = Array(o.length);
    const u = [];
    for (let c = 0; c < o.length; c++) {
      const a = o[c], l = r.element._zod.run({
        value: a,
        issues: []
      }, t);
      l instanceof Promise ? u.push(l.then((p) => Lt(p, n, c))) : Lt(l, n, c);
    }
    return u.length ? Promise.all(u).then(() => n) : n;
  };
});
function Xr(e, r, n, t) {
  e.issues.length && r.issues.push(...$o(n, e.issues)), e.value === void 0 ? n in t && (r.value[n] = void 0) : r.value[n] = e.value;
}
function So(e) {
  const r = Object.keys(e.shape);
  for (const t of r)
    if (!e.shape?.[t]?._zod?.traits?.has("$ZodType"))
      throw new Error(`Invalid element at key "${t}": expected a Zod schema`);
  const n = ru(e.shape);
  return {
    ...e,
    keys: r,
    keySet: new Set(r),
    numKeys: r.length,
    optionalKeys: new Set(n)
  };
}
function Zo(e, r, n, t, o, u) {
  const c = [], a = o.keySet, l = o.catchall._zod, p = l.def.type;
  for (const b of Object.keys(r)) {
    if (a.has(b))
      continue;
    if (p === "never") {
      c.push(b);
      continue;
    }
    const y = l.run({ value: r[b], issues: [] }, t);
    y instanceof Promise ? e.push(y.then((O) => Xr(O, n, b, r))) : Xr(y, n, b, r);
  }
  return c.length && n.issues.push({
    code: "unrecognized_keys",
    keys: c,
    input: r,
    inst: u
  }), e.length ? Promise.all(e).then(() => n) : n;
}
const Uc = /* @__PURE__ */ f("$ZodObject", (e, r) => {
  if (z.init(e, r), !Object.getOwnPropertyDescriptor(r, "shape")?.get) {
    const a = r.shape;
    Object.defineProperty(r, "shape", {
      get: () => {
        const l = { ...a };
        return Object.defineProperty(r, "shape", {
          value: l
        }), l;
      }
    });
  }
  const t = Vn(() => So(r));
  k(e._zod, "propValues", () => {
    const a = r.shape, l = {};
    for (const p in a) {
      const b = a[p]._zod;
      if (b.values) {
        l[p] ?? (l[p] = /* @__PURE__ */ new Set());
        for (const y of b.values)
          l[p].add(y);
      }
    }
    return l;
  });
  const o = Yr, u = r.catchall;
  let c;
  e._zod.parse = (a, l) => {
    c ?? (c = t.value);
    const p = a.value;
    if (!o(p))
      return a.issues.push({
        expected: "object",
        code: "invalid_type",
        input: p,
        inst: e
      }), a;
    a.value = {};
    const b = [], y = c.shape;
    for (const O of c.keys) {
      const Q = y[O]._zod.run({ value: p[O], issues: [] }, l);
      Q instanceof Promise ? b.push(Q.then((zn) => Xr(zn, a, O, p))) : Xr(Q, a, O, p);
    }
    return u ? Zo(b, p, a, l, t.value, e) : b.length ? Promise.all(b).then(() => a) : a;
  };
}), Dc = /* @__PURE__ */ f("$ZodObjectJIT", (e, r) => {
  Uc.init(e, r);
  const n = e._zod.parse, t = Vn(() => So(r)), o = (O) => {
    const I = new cc(["shape", "payload", "ctx"]), Q = t.value, zn = (Fe) => {
      const ge = Ut(Fe);
      return `shape[${ge}]._zod.run({ value: input[${ge}], issues: [] }, ctx)`;
    };
    I.write("const input = payload.value;");
    const Et = /* @__PURE__ */ Object.create(null);
    let Zi = 0;
    for (const Fe of Q.keys)
      Et[Fe] = `key_${Zi++}`;
    I.write("const newResult = {};");
    for (const Fe of Q.keys) {
      const ge = Et[Fe], Fr = Ut(Fe);
      I.write(`const ${ge} = ${zn(Fe)};`), I.write(`
        if (${ge}.issues.length) {
          payload.issues = payload.issues.concat(${ge}.issues.map(iss => ({
            ...iss,
            path: iss.path ? [${Fr}, ...iss.path] : [${Fr}]
          })));
        }
        
        
        if (${ge}.value === undefined) {
          if (${Fr} in input) {
            newResult[${Fr}] = undefined;
          }
        } else {
          newResult[${Fr}] = ${ge}.value;
        }
        
      `);
    }
    I.write("payload.value = newResult;"), I.write("return payload;");
    const ji = I.compile();
    return (Fe, ge) => ji(O, Fe, ge);
  };
  let u;
  const c = Yr, a = !Fo.jitless, p = a && Qi.value, b = r.catchall;
  let y;
  e._zod.parse = (O, I) => {
    y ?? (y = t.value);
    const Q = O.value;
    return c(Q) ? a && p && I?.async === !1 && I.jitless !== !0 ? (u || (u = o(r.shape)), O = u(O, I), b ? Zo([], Q, O, I, y, e) : O) : n(O, I) : (O.issues.push({
      expected: "object",
      code: "invalid_type",
      input: Q,
      inst: e
    }), O);
  };
});
function Mt(e, r, n, t) {
  for (const u of e)
    if (u.issues.length === 0)
      return r.value = u.value, r;
  const o = e.filter((u) => !er(u));
  return o.length === 1 ? (r.value = o[0].value, o[0]) : (r.issues.push({
    code: "invalid_union",
    input: r.value,
    inst: n,
    errors: e.map((u) => u.issues.map((c) => Ue(c, t, Ne())))
  }), r);
}
const Lc = /* @__PURE__ */ f("$ZodUnion", (e, r) => {
  z.init(e, r), k(e._zod, "optin", () => r.options.some((o) => o._zod.optin === "optional") ? "optional" : void 0), k(e._zod, "optout", () => r.options.some((o) => o._zod.optout === "optional") ? "optional" : void 0), k(e._zod, "values", () => {
    if (r.options.every((o) => o._zod.values))
      return new Set(r.options.flatMap((o) => Array.from(o._zod.values)));
  }), k(e._zod, "pattern", () => {
    if (r.options.every((o) => o._zod.pattern)) {
      const o = r.options.map((u) => u._zod.pattern);
      return new RegExp(`^(${o.map((u) => Gn(u.source)).join("|")})$`);
    }
  });
  const n = r.options.length === 1, t = r.options[0]._zod.run;
  e._zod.parse = (o, u) => {
    if (n)
      return t(o, u);
    let c = !1;
    const a = [];
    for (const l of r.options) {
      const p = l._zod.run({
        value: o.value,
        issues: []
      }, u);
      if (p instanceof Promise)
        a.push(p), c = !0;
      else {
        if (p.issues.length === 0)
          return p;
        a.push(p);
      }
    }
    return c ? Promise.all(a).then((l) => Mt(l, o, e, u)) : Mt(a, o, e, u);
  };
}), Mc = /* @__PURE__ */ f("$ZodIntersection", (e, r) => {
  z.init(e, r), e._zod.parse = (n, t) => {
    const o = n.value, u = r.left._zod.run({ value: o, issues: [] }, t), c = r.right._zod.run({ value: o, issues: [] }, t);
    return u instanceof Promise || c instanceof Promise ? Promise.all([u, c]).then(([l, p]) => Wt(n, l, p)) : Wt(n, u, c);
  };
});
function Cn(e, r) {
  if (e === r)
    return { valid: !0, data: e };
  if (e instanceof Date && r instanceof Date && +e == +r)
    return { valid: !0, data: e };
  if (wr(e) && wr(r)) {
    const n = Object.keys(r), t = Object.keys(e).filter((u) => n.indexOf(u) !== -1), o = { ...e, ...r };
    for (const u of t) {
      const c = Cn(e[u], r[u]);
      if (!c.valid)
        return {
          valid: !1,
          mergeErrorPath: [u, ...c.mergeErrorPath]
        };
      o[u] = c.data;
    }
    return { valid: !0, data: o };
  }
  if (Array.isArray(e) && Array.isArray(r)) {
    if (e.length !== r.length)
      return { valid: !1, mergeErrorPath: [] };
    const n = [];
    for (let t = 0; t < e.length; t++) {
      const o = e[t], u = r[t], c = Cn(o, u);
      if (!c.valid)
        return {
          valid: !1,
          mergeErrorPath: [t, ...c.mergeErrorPath]
        };
      n.push(c.data);
    }
    return { valid: !0, data: n };
  }
  return { valid: !1, mergeErrorPath: [] };
}
function Wt(e, r, n) {
  if (r.issues.length && e.issues.push(...r.issues), n.issues.length && e.issues.push(...n.issues), er(e))
    return e;
  const t = Cn(r.value, n.value);
  if (!t.valid)
    throw new Error(`Unmergable intersection. Error path: ${JSON.stringify(t.mergeErrorPath)}`);
  return e.value = t.data, e;
}
const Wc = /* @__PURE__ */ f("$ZodEnum", (e, r) => {
  z.init(e, r);
  const n = Yi(r.entries), t = new Set(n);
  e._zod.values = t, e._zod.pattern = new RegExp(`^(${n.filter((o) => eu.has(typeof o)).map((o) => typeof o == "string" ? tr(o) : o.toString()).join("|")})$`), e._zod.parse = (o, u) => {
    const c = o.value;
    return t.has(c) || o.issues.push({
      code: "invalid_value",
      values: n,
      input: c,
      inst: e
    }), o;
  };
}), Bc = /* @__PURE__ */ f("$ZodLiteral", (e, r) => {
  if (z.init(e, r), r.values.length === 0)
    throw new Error("Cannot create literal schema with no valid values");
  e._zod.values = new Set(r.values), e._zod.pattern = new RegExp(`^(${r.values.map((n) => typeof n == "string" ? tr(n) : n ? tr(n.toString()) : String(n)).join("|")})$`), e._zod.parse = (n, t) => {
    const o = n.value;
    return e._zod.values.has(o) || n.issues.push({
      code: "invalid_value",
      values: r.values,
      input: o,
      inst: e
    }), n;
  };
}), Kc = /* @__PURE__ */ f("$ZodTransform", (e, r) => {
  z.init(e, r), e._zod.parse = (n, t) => {
    if (t.direction === "backward")
      throw new bo(e.constructor.name);
    const o = r.transform(n.value, n);
    if (t.async)
      return (o instanceof Promise ? o : Promise.resolve(o)).then((c) => (n.value = c, n));
    if (o instanceof Promise)
      throw new rr();
    return n.value = o, n;
  };
});
function Bt(e, r) {
  return e.issues.length && r === void 0 ? { issues: [], value: void 0 } : e;
}
const Vc = /* @__PURE__ */ f("$ZodOptional", (e, r) => {
  z.init(e, r), e._zod.optin = "optional", e._zod.optout = "optional", k(e._zod, "values", () => r.innerType._zod.values ? /* @__PURE__ */ new Set([...r.innerType._zod.values, void 0]) : void 0), k(e._zod, "pattern", () => {
    const n = r.innerType._zod.pattern;
    return n ? new RegExp(`^(${Gn(n.source)})?$`) : void 0;
  }), e._zod.parse = (n, t) => {
    if (r.innerType._zod.optin === "optional") {
      const o = r.innerType._zod.run(n, t);
      return o instanceof Promise ? o.then((u) => Bt(u, n.value)) : Bt(o, n.value);
    }
    return n.value === void 0 ? n : r.innerType._zod.run(n, t);
  };
}), Hc = /* @__PURE__ */ f("$ZodNullable", (e, r) => {
  z.init(e, r), k(e._zod, "optin", () => r.innerType._zod.optin), k(e._zod, "optout", () => r.innerType._zod.optout), k(e._zod, "pattern", () => {
    const n = r.innerType._zod.pattern;
    return n ? new RegExp(`^(${Gn(n.source)}|null)$`) : void 0;
  }), k(e._zod, "values", () => r.innerType._zod.values ? /* @__PURE__ */ new Set([...r.innerType._zod.values, null]) : void 0), e._zod.parse = (n, t) => n.value === null ? n : r.innerType._zod.run(n, t);
}), Gc = /* @__PURE__ */ f("$ZodDefault", (e, r) => {
  z.init(e, r), e._zod.optin = "optional", k(e._zod, "values", () => r.innerType._zod.values), e._zod.parse = (n, t) => {
    if (t.direction === "backward")
      return r.innerType._zod.run(n, t);
    if (n.value === void 0)
      return n.value = r.defaultValue, n;
    const o = r.innerType._zod.run(n, t);
    return o instanceof Promise ? o.then((u) => Kt(u, r)) : Kt(o, r);
  };
});
function Kt(e, r) {
  return e.value === void 0 && (e.value = r.defaultValue), e;
}
const qc = /* @__PURE__ */ f("$ZodPrefault", (e, r) => {
  z.init(e, r), e._zod.optin = "optional", k(e._zod, "values", () => r.innerType._zod.values), e._zod.parse = (n, t) => (t.direction === "backward" || n.value === void 0 && (n.value = r.defaultValue), r.innerType._zod.run(n, t));
}), Jc = /* @__PURE__ */ f("$ZodNonOptional", (e, r) => {
  z.init(e, r), k(e._zod, "values", () => {
    const n = r.innerType._zod.values;
    return n ? new Set([...n].filter((t) => t !== void 0)) : void 0;
  }), e._zod.parse = (n, t) => {
    const o = r.innerType._zod.run(n, t);
    return o instanceof Promise ? o.then((u) => Vt(u, e)) : Vt(o, e);
  };
});
function Vt(e, r) {
  return !e.issues.length && e.value === void 0 && e.issues.push({
    code: "invalid_type",
    expected: "nonoptional",
    input: e.value,
    inst: r
  }), e;
}
const Yc = /* @__PURE__ */ f("$ZodCatch", (e, r) => {
  z.init(e, r), k(e._zod, "optin", () => r.innerType._zod.optin), k(e._zod, "optout", () => r.innerType._zod.optout), k(e._zod, "values", () => r.innerType._zod.values), e._zod.parse = (n, t) => {
    if (t.direction === "backward")
      return r.innerType._zod.run(n, t);
    const o = r.innerType._zod.run(n, t);
    return o instanceof Promise ? o.then((u) => (n.value = u.value, u.issues.length && (n.value = r.catchValue({
      ...n,
      error: {
        issues: u.issues.map((c) => Ue(c, t, Ne()))
      },
      input: n.value
    }), n.issues = []), n)) : (n.value = o.value, o.issues.length && (n.value = r.catchValue({
      ...n,
      error: {
        issues: o.issues.map((u) => Ue(u, t, Ne()))
      },
      input: n.value
    }), n.issues = []), n);
  };
}), Xc = /* @__PURE__ */ f("$ZodPipe", (e, r) => {
  z.init(e, r), k(e._zod, "values", () => r.in._zod.values), k(e._zod, "optin", () => r.in._zod.optin), k(e._zod, "optout", () => r.out._zod.optout), k(e._zod, "propValues", () => r.in._zod.propValues), e._zod.parse = (n, t) => {
    if (t.direction === "backward") {
      const u = r.out._zod.run(n, t);
      return u instanceof Promise ? u.then((c) => Vr(c, r.in, t)) : Vr(u, r.in, t);
    }
    const o = r.in._zod.run(n, t);
    return o instanceof Promise ? o.then((u) => Vr(u, r.out, t)) : Vr(o, r.out, t);
  };
});
function Vr(e, r, n) {
  return e.issues.length ? (e.aborted = !0, e) : r._zod.run({ value: e.value, issues: e.issues }, n);
}
const Qc = /* @__PURE__ */ f("$ZodReadonly", (e, r) => {
  z.init(e, r), k(e._zod, "propValues", () => r.innerType._zod.propValues), k(e._zod, "values", () => r.innerType._zod.values), k(e._zod, "optin", () => r.innerType._zod.optin), k(e._zod, "optout", () => r.innerType._zod.optout), e._zod.parse = (n, t) => {
    if (t.direction === "backward")
      return r.innerType._zod.run(n, t);
    const o = r.innerType._zod.run(n, t);
    return o instanceof Promise ? o.then(Ht) : Ht(o);
  };
});
function Ht(e) {
  return e.value = Object.freeze(e.value), e;
}
const ea = /* @__PURE__ */ f("$ZodCustom", (e, r) => {
  Y.init(e, r), z.init(e, r), e._zod.parse = (n, t) => n, e._zod.check = (n) => {
    const t = n.value, o = r.fn(t);
    if (o instanceof Promise)
      return o.then((u) => Gt(u, n, t, e));
    Gt(o, n, t, e);
  };
});
function Gt(e, r, n, t) {
  if (!e) {
    const o = {
      code: "custom",
      input: n,
      inst: t,
      // incorporates params.error into issue reporting
      path: [...t._zod.def.path ?? []],
      // incorporates params.error into issue reporting
      continue: !t._zod.def.abort
      // params: inst._zod.def.params,
    };
    t._zod.def.params && (o.params = t._zod.def.params), r.issues.push(yr(o));
  }
}
class ra {
  constructor() {
    this._map = /* @__PURE__ */ new WeakMap(), this._idmap = /* @__PURE__ */ new Map();
  }
  add(r, ...n) {
    const t = n[0];
    if (this._map.set(r, t), t && typeof t == "object" && "id" in t) {
      if (this._idmap.has(t.id))
        throw new Error(`ID ${t.id} already exists in the registry`);
      this._idmap.set(t.id, r);
    }
    return this;
  }
  clear() {
    return this._map = /* @__PURE__ */ new WeakMap(), this._idmap = /* @__PURE__ */ new Map(), this;
  }
  remove(r) {
    const n = this._map.get(r);
    return n && typeof n == "object" && "id" in n && this._idmap.delete(n.id), this._map.delete(r), this;
  }
  get(r) {
    const n = r._zod.parent;
    if (n) {
      const t = { ...this.get(n) ?? {} };
      delete t.id;
      const o = { ...t, ...this._map.get(r) };
      return Object.keys(o).length ? o : void 0;
    }
    return this._map.get(r);
  }
  has(r) {
    return this._map.has(r);
  }
}
function na() {
  return new ra();
}
const Hr = /* @__PURE__ */ na();
function ta(e, r) {
  return new e({
    type: "string",
    ...h(r)
  });
}
function oa(e, r) {
  return new e({
    type: "string",
    format: "email",
    check: "string_format",
    abort: !1,
    ...h(r)
  });
}
function qt(e, r) {
  return new e({
    type: "string",
    format: "guid",
    check: "string_format",
    abort: !1,
    ...h(r)
  });
}
function ia(e, r) {
  return new e({
    type: "string",
    format: "uuid",
    check: "string_format",
    abort: !1,
    ...h(r)
  });
}
function ua(e, r) {
  return new e({
    type: "string",
    format: "uuid",
    check: "string_format",
    abort: !1,
    version: "v4",
    ...h(r)
  });
}
function ca(e, r) {
  return new e({
    type: "string",
    format: "uuid",
    check: "string_format",
    abort: !1,
    version: "v6",
    ...h(r)
  });
}
function aa(e, r) {
  return new e({
    type: "string",
    format: "uuid",
    check: "string_format",
    abort: !1,
    version: "v7",
    ...h(r)
  });
}
function sa(e, r) {
  return new e({
    type: "string",
    format: "url",
    check: "string_format",
    abort: !1,
    ...h(r)
  });
}
function fa(e, r) {
  return new e({
    type: "string",
    format: "emoji",
    check: "string_format",
    abort: !1,
    ...h(r)
  });
}
function da(e, r) {
  return new e({
    type: "string",
    format: "nanoid",
    check: "string_format",
    abort: !1,
    ...h(r)
  });
}
function ma(e, r) {
  return new e({
    type: "string",
    format: "cuid",
    check: "string_format",
    abort: !1,
    ...h(r)
  });
}
function la(e, r) {
  return new e({
    type: "string",
    format: "cuid2",
    check: "string_format",
    abort: !1,
    ...h(r)
  });
}
function pa(e, r) {
  return new e({
    type: "string",
    format: "ulid",
    check: "string_format",
    abort: !1,
    ...h(r)
  });
}
function ha(e, r) {
  return new e({
    type: "string",
    format: "xid",
    check: "string_format",
    abort: !1,
    ...h(r)
  });
}
function ga(e, r) {
  return new e({
    type: "string",
    format: "ksuid",
    check: "string_format",
    abort: !1,
    ...h(r)
  });
}
function ba(e, r) {
  return new e({
    type: "string",
    format: "ipv4",
    check: "string_format",
    abort: !1,
    ...h(r)
  });
}
function Fa(e, r) {
  return new e({
    type: "string",
    format: "ipv6",
    check: "string_format",
    abort: !1,
    ...h(r)
  });
}
function va(e, r) {
  return new e({
    type: "string",
    format: "cidrv4",
    check: "string_format",
    abort: !1,
    ...h(r)
  });
}
function Oa(e, r) {
  return new e({
    type: "string",
    format: "cidrv6",
    check: "string_format",
    abort: !1,
    ...h(r)
  });
}
function $a(e, r) {
  return new e({
    type: "string",
    format: "base64",
    check: "string_format",
    abort: !1,
    ...h(r)
  });
}
function _a(e, r) {
  return new e({
    type: "string",
    format: "base64url",
    check: "string_format",
    abort: !1,
    ...h(r)
  });
}
function wa(e, r) {
  return new e({
    type: "string",
    format: "e164",
    check: "string_format",
    abort: !1,
    ...h(r)
  });
}
function ya(e, r) {
  return new e({
    type: "string",
    format: "jwt",
    check: "string_format",
    abort: !1,
    ...h(r)
  });
}
function ka(e, r) {
  return new e({
    type: "string",
    format: "datetime",
    check: "string_format",
    offset: !1,
    local: !1,
    precision: null,
    ...h(r)
  });
}
function Ia(e, r) {
  return new e({
    type: "string",
    format: "date",
    check: "string_format",
    ...h(r)
  });
}
function Ra(e, r) {
  return new e({
    type: "string",
    format: "time",
    check: "string_format",
    precision: null,
    ...h(r)
  });
}
function xa(e, r) {
  return new e({
    type: "string",
    format: "duration",
    check: "string_format",
    ...h(r)
  });
}
function Aa(e, r) {
  return new e({
    type: "number",
    checks: [],
    ...h(r)
  });
}
function za(e, r) {
  return new e({
    type: "number",
    check: "number_format",
    abort: !1,
    format: "safeint",
    ...h(r)
  });
}
function Ta(e, r) {
  return new e({
    type: "boolean",
    ...h(r)
  });
}
function Pa(e) {
  return new e({
    type: "unknown"
  });
}
function Sa(e, r) {
  return new e({
    type: "never",
    ...h(r)
  });
}
function Jt(e, r) {
  return new Ao({
    check: "less_than",
    ...h(r),
    value: e,
    inclusive: !1
  });
}
function Pn(e, r) {
  return new Ao({
    check: "less_than",
    ...h(r),
    value: e,
    inclusive: !0
  });
}
function Yt(e, r) {
  return new zo({
    check: "greater_than",
    ...h(r),
    value: e,
    inclusive: !1
  });
}
function Sn(e, r) {
  return new zo({
    check: "greater_than",
    ...h(r),
    value: e,
    inclusive: !0
  });
}
function Xt(e, r) {
  return new qu({
    check: "multiple_of",
    ...h(r),
    value: e
  });
}
function jo(e, r) {
  return new Yu({
    check: "max_length",
    ...h(r),
    maximum: e
  });
}
function Qr(e, r) {
  return new Xu({
    check: "min_length",
    ...h(r),
    minimum: e
  });
}
function Eo(e, r) {
  return new Qu({
    check: "length_equals",
    ...h(r),
    length: e
  });
}
function Za(e, r) {
  return new ec({
    check: "string_format",
    format: "regex",
    ...h(r),
    pattern: e
  });
}
function ja(e) {
  return new rc({
    check: "string_format",
    format: "lowercase",
    ...h(e)
  });
}
function Ea(e) {
  return new nc({
    check: "string_format",
    format: "uppercase",
    ...h(e)
  });
}
function Ca(e, r) {
  return new tc({
    check: "string_format",
    format: "includes",
    ...h(r),
    includes: e
  });
}
function Na(e, r) {
  return new oc({
    check: "string_format",
    format: "starts_with",
    ...h(r),
    prefix: e
  });
}
function Ua(e, r) {
  return new ic({
    check: "string_format",
    format: "ends_with",
    ...h(r),
    suffix: e
  });
}
function Pr(e) {
  return new uc({
    check: "overwrite",
    tx: e
  });
}
function Da(e) {
  return Pr((r) => r.normalize(e));
}
function La() {
  return Pr((e) => e.trim());
}
function Ma() {
  return Pr((e) => e.toLowerCase());
}
function Wa() {
  return Pr((e) => e.toUpperCase());
}
function Ba(e, r, n) {
  return new e({
    type: "array",
    element: r,
    // get element() {
    //   return element;
    // },
    ...h(n)
  });
}
function Ka(e, r, n) {
  return new e({
    type: "custom",
    check: "custom",
    fn: r,
    ...h(n)
  });
}
function Va(e) {
  const r = Ha((n) => (n.addIssue = (t) => {
    if (typeof t == "string")
      n.issues.push(yr(t, n.value, r._zod.def));
    else {
      const o = t;
      o.fatal && (o.continue = !1), o.code ?? (o.code = "custom"), o.input ?? (o.input = n.value), o.inst ?? (o.inst = r), o.continue ?? (o.continue = !r._zod.def.abort), n.issues.push(yr(o));
    }
  }, e(n.value, n)));
  return r;
}
function Ha(e, r) {
  const n = new Y({
    check: "custom",
    ...h(r)
  });
  return n._zod.check = e, n;
}
const Ga = /* @__PURE__ */ f("ZodISODateTime", (e, r) => {
  Oc.init(e, r), x.init(e, r);
});
function qa(e) {
  return ka(Ga, e);
}
const Ja = /* @__PURE__ */ f("ZodISODate", (e, r) => {
  $c.init(e, r), x.init(e, r);
});
function Ya(e) {
  return Ia(Ja, e);
}
const Xa = /* @__PURE__ */ f("ZodISOTime", (e, r) => {
  _c.init(e, r), x.init(e, r);
});
function Qa(e) {
  return Ra(Xa, e);
}
const es = /* @__PURE__ */ f("ZodISODuration", (e, r) => {
  wc.init(e, r), x.init(e, r);
});
function rs(e) {
  return xa(es, e);
}
const ns = (e, r) => {
  wo.init(e, r), e.name = "ZodError", Object.defineProperties(e, {
    format: {
      value: (n) => du(e, n)
      // enumerable: false,
    },
    flatten: {
      value: (n) => fu(e, n)
      // enumerable: false,
    },
    addIssue: {
      value: (n) => {
        e.issues.push(n), e.message = JSON.stringify(e.issues, En, 2);
      }
      // enumerable: false,
    },
    addIssues: {
      value: (n) => {
        e.issues.push(...n), e.message = JSON.stringify(e.issues, En, 2);
      }
      // enumerable: false,
    },
    isEmpty: {
      get() {
        return e.issues.length === 0;
      }
      // enumerable: false,
    }
  });
}, ae = f("ZodError", ns, {
  Parent: Error
}), ts = /* @__PURE__ */ Jn(ae), os = /* @__PURE__ */ Yn(ae), is = /* @__PURE__ */ on(ae), us = /* @__PURE__ */ un(ae), cs = /* @__PURE__ */ pu(ae), as = /* @__PURE__ */ hu(ae), ss = /* @__PURE__ */ gu(ae), fs = /* @__PURE__ */ bu(ae), ds = /* @__PURE__ */ Fu(ae), ms = /* @__PURE__ */ vu(ae), ls = /* @__PURE__ */ Ou(ae), ps = /* @__PURE__ */ $u(ae), P = /* @__PURE__ */ f("ZodType", (e, r) => (z.init(e, r), e.def = r, e.type = r.type, Object.defineProperty(e, "_def", { value: r }), e.check = (...n) => e.clone(We(r, {
  checks: [
    ...r.checks ?? [],
    ...n.map((t) => typeof t == "function" ? { _zod: { check: t, def: { check: "custom" }, onattach: [] } } : t)
  ]
})), e.clone = (n, t) => xe(e, n, t), e.brand = () => e, e.register = ((n, t) => (n.add(e, t), e)), e.parse = (n, t) => ts(e, n, t, { callee: e.parse }), e.safeParse = (n, t) => is(e, n, t), e.parseAsync = async (n, t) => os(e, n, t, { callee: e.parseAsync }), e.safeParseAsync = async (n, t) => us(e, n, t), e.spa = e.safeParseAsync, e.encode = (n, t) => cs(e, n, t), e.decode = (n, t) => as(e, n, t), e.encodeAsync = async (n, t) => ss(e, n, t), e.decodeAsync = async (n, t) => fs(e, n, t), e.safeEncode = (n, t) => ds(e, n, t), e.safeDecode = (n, t) => ms(e, n, t), e.safeEncodeAsync = async (n, t) => ls(e, n, t), e.safeDecodeAsync = async (n, t) => ps(e, n, t), e.refine = (n, t) => e.check(af(n, t)), e.superRefine = (n) => e.check(sf(n)), e.overwrite = (n) => e.check(Pr(n)), e.optional = () => nr(e), e.nullable = () => no(e), e.nullish = () => nr(no(e)), e.nonoptional = (n) => ef(e, n), e.array = () => Nn(e), e.or = (n) => Qn([e, n]), e.and = (n) => Bs(e, n), e.transform = (n) => to(e, Gs(n)), e.default = (n) => Ys(e, n), e.prefault = (n) => Qs(e, n), e.catch = (n) => nf(e, n), e.pipe = (n) => to(e, n), e.readonly = () => uf(e), e.describe = (n) => {
  const t = e.clone();
  return Hr.add(t, { description: n }), t;
}, Object.defineProperty(e, "description", {
  get() {
    return Hr.get(e)?.description;
  },
  configurable: !0
}), e.meta = (...n) => {
  if (n.length === 0)
    return Hr.get(e);
  const t = e.clone();
  return Hr.add(t, n[0]), t;
}, e.isOptional = () => e.safeParse(void 0).success, e.isNullable = () => e.safeParse(null).success, e)), Co = /* @__PURE__ */ f("_ZodString", (e, r) => {
  Xn.init(e, r), P.init(e, r);
  const n = e._zod.bag;
  e.format = n.format ?? null, e.minLength = n.minimum ?? null, e.maxLength = n.maximum ?? null, e.regex = (...t) => e.check(Za(...t)), e.includes = (...t) => e.check(Ca(...t)), e.startsWith = (...t) => e.check(Na(...t)), e.endsWith = (...t) => e.check(Ua(...t)), e.min = (...t) => e.check(Qr(...t)), e.max = (...t) => e.check(jo(...t)), e.length = (...t) => e.check(Eo(...t)), e.nonempty = (...t) => e.check(Qr(1, ...t)), e.lowercase = (t) => e.check(ja(t)), e.uppercase = (t) => e.check(Ea(t)), e.trim = () => e.check(La()), e.normalize = (...t) => e.check(Da(...t)), e.toLowerCase = () => e.check(Ma()), e.toUpperCase = () => e.check(Wa());
}), hs = /* @__PURE__ */ f("ZodString", (e, r) => {
  Xn.init(e, r), Co.init(e, r), e.email = (n) => e.check(oa(gs, n)), e.url = (n) => e.check(sa(bs, n)), e.jwt = (n) => e.check(ya(Ps, n)), e.emoji = (n) => e.check(fa(Fs, n)), e.guid = (n) => e.check(qt(Qt, n)), e.uuid = (n) => e.check(ia(Gr, n)), e.uuidv4 = (n) => e.check(ua(Gr, n)), e.uuidv6 = (n) => e.check(ca(Gr, n)), e.uuidv7 = (n) => e.check(aa(Gr, n)), e.nanoid = (n) => e.check(da(vs, n)), e.guid = (n) => e.check(qt(Qt, n)), e.cuid = (n) => e.check(ma(Os, n)), e.cuid2 = (n) => e.check(la($s, n)), e.ulid = (n) => e.check(pa(_s, n)), e.base64 = (n) => e.check($a(As, n)), e.base64url = (n) => e.check(_a(zs, n)), e.xid = (n) => e.check(ha(ws, n)), e.ksuid = (n) => e.check(ga(ys, n)), e.ipv4 = (n) => e.check(ba(ks, n)), e.ipv6 = (n) => e.check(Fa(Is, n)), e.cidrv4 = (n) => e.check(va(Rs, n)), e.cidrv6 = (n) => e.check(Oa(xs, n)), e.e164 = (n) => e.check(wa(Ts, n)), e.datetime = (n) => e.check(qa(n)), e.date = (n) => e.check(Ya(n)), e.time = (n) => e.check(Qa(n)), e.duration = (n) => e.check(rs(n));
});
function kr(e) {
  return ta(hs, e);
}
const x = /* @__PURE__ */ f("ZodStringFormat", (e, r) => {
  R.init(e, r), Co.init(e, r);
}), gs = /* @__PURE__ */ f("ZodEmail", (e, r) => {
  dc.init(e, r), x.init(e, r);
}), Qt = /* @__PURE__ */ f("ZodGUID", (e, r) => {
  sc.init(e, r), x.init(e, r);
}), Gr = /* @__PURE__ */ f("ZodUUID", (e, r) => {
  fc.init(e, r), x.init(e, r);
}), bs = /* @__PURE__ */ f("ZodURL", (e, r) => {
  mc.init(e, r), x.init(e, r);
}), Fs = /* @__PURE__ */ f("ZodEmoji", (e, r) => {
  lc.init(e, r), x.init(e, r);
}), vs = /* @__PURE__ */ f("ZodNanoID", (e, r) => {
  pc.init(e, r), x.init(e, r);
}), Os = /* @__PURE__ */ f("ZodCUID", (e, r) => {
  hc.init(e, r), x.init(e, r);
}), $s = /* @__PURE__ */ f("ZodCUID2", (e, r) => {
  gc.init(e, r), x.init(e, r);
}), _s = /* @__PURE__ */ f("ZodULID", (e, r) => {
  bc.init(e, r), x.init(e, r);
}), ws = /* @__PURE__ */ f("ZodXID", (e, r) => {
  Fc.init(e, r), x.init(e, r);
}), ys = /* @__PURE__ */ f("ZodKSUID", (e, r) => {
  vc.init(e, r), x.init(e, r);
}), ks = /* @__PURE__ */ f("ZodIPv4", (e, r) => {
  yc.init(e, r), x.init(e, r);
}), Is = /* @__PURE__ */ f("ZodIPv6", (e, r) => {
  kc.init(e, r), x.init(e, r);
}), Rs = /* @__PURE__ */ f("ZodCIDRv4", (e, r) => {
  Ic.init(e, r), x.init(e, r);
}), xs = /* @__PURE__ */ f("ZodCIDRv6", (e, r) => {
  Rc.init(e, r), x.init(e, r);
}), As = /* @__PURE__ */ f("ZodBase64", (e, r) => {
  xc.init(e, r), x.init(e, r);
}), zs = /* @__PURE__ */ f("ZodBase64URL", (e, r) => {
  zc.init(e, r), x.init(e, r);
}), Ts = /* @__PURE__ */ f("ZodE164", (e, r) => {
  Tc.init(e, r), x.init(e, r);
}), Ps = /* @__PURE__ */ f("ZodJWT", (e, r) => {
  Sc.init(e, r), x.init(e, r);
}), No = /* @__PURE__ */ f("ZodNumber", (e, r) => {
  Po.init(e, r), P.init(e, r), e.gt = (t, o) => e.check(Yt(t, o)), e.gte = (t, o) => e.check(Sn(t, o)), e.min = (t, o) => e.check(Sn(t, o)), e.lt = (t, o) => e.check(Jt(t, o)), e.lte = (t, o) => e.check(Pn(t, o)), e.max = (t, o) => e.check(Pn(t, o)), e.int = (t) => e.check(eo(t)), e.safe = (t) => e.check(eo(t)), e.positive = (t) => e.check(Yt(0, t)), e.nonnegative = (t) => e.check(Sn(0, t)), e.negative = (t) => e.check(Jt(0, t)), e.nonpositive = (t) => e.check(Pn(0, t)), e.multipleOf = (t, o) => e.check(Xt(t, o)), e.step = (t, o) => e.check(Xt(t, o)), e.finite = () => e;
  const n = e._zod.bag;
  e.minValue = Math.max(n.minimum ?? Number.NEGATIVE_INFINITY, n.exclusiveMinimum ?? Number.NEGATIVE_INFINITY) ?? null, e.maxValue = Math.min(n.maximum ?? Number.POSITIVE_INFINITY, n.exclusiveMaximum ?? Number.POSITIVE_INFINITY) ?? null, e.isInt = (n.format ?? "").includes("int") || Number.isSafeInteger(n.multipleOf ?? 0.5), e.isFinite = !0, e.format = n.format ?? null;
});
function Ss(e) {
  return Aa(No, e);
}
const Zs = /* @__PURE__ */ f("ZodNumberFormat", (e, r) => {
  Zc.init(e, r), No.init(e, r);
});
function eo(e) {
  return za(Zs, e);
}
const js = /* @__PURE__ */ f("ZodBoolean", (e, r) => {
  jc.init(e, r), P.init(e, r);
});
function Es(e) {
  return Ta(js, e);
}
const Cs = /* @__PURE__ */ f("ZodUnknown", (e, r) => {
  Ec.init(e, r), P.init(e, r);
});
function ro() {
  return Pa(Cs);
}
const Ns = /* @__PURE__ */ f("ZodNever", (e, r) => {
  Cc.init(e, r), P.init(e, r);
});
function Us(e) {
  return Sa(Ns, e);
}
const Ds = /* @__PURE__ */ f("ZodArray", (e, r) => {
  Nc.init(e, r), P.init(e, r), e.element = r.element, e.min = (n, t) => e.check(Qr(n, t)), e.nonempty = (n) => e.check(Qr(1, n)), e.max = (n, t) => e.check(jo(n, t)), e.length = (n, t) => e.check(Eo(n, t)), e.unwrap = () => e.element;
});
function Nn(e, r) {
  return Ba(Ds, e, r);
}
const Ls = /* @__PURE__ */ f("ZodObject", (e, r) => {
  Dc.init(e, r), P.init(e, r), k(e, "shape", () => r.shape), e.keyof = () => Ks(Object.keys(e._zod.def.shape)), e.catchall = (n) => e.clone({ ...e._zod.def, catchall: n }), e.passthrough = () => e.clone({ ...e._zod.def, catchall: ro() }), e.loose = () => e.clone({ ...e._zod.def, catchall: ro() }), e.strict = () => e.clone({ ...e._zod.def, catchall: Us() }), e.strip = () => e.clone({ ...e._zod.def, catchall: void 0 }), e.extend = (n) => iu(e, n), e.safeExtend = (n) => uu(e, n), e.merge = (n) => cu(e, n), e.pick = (n) => tu(e, n), e.omit = (n) => ou(e, n), e.partial = (...n) => au(Uo, e, n[0]), e.required = (...n) => su(Do, e, n[0]);
});
function Ee(e, r) {
  const n = {
    type: "object",
    shape: e ?? {},
    ...h(r)
  };
  return new Ls(n);
}
const Ms = /* @__PURE__ */ f("ZodUnion", (e, r) => {
  Lc.init(e, r), P.init(e, r), e.options = r.options;
});
function Qn(e, r) {
  return new Ms({
    type: "union",
    options: e,
    ...h(r)
  });
}
const Ws = /* @__PURE__ */ f("ZodIntersection", (e, r) => {
  Mc.init(e, r), P.init(e, r);
});
function Bs(e, r) {
  return new Ws({
    type: "intersection",
    left: e,
    right: r
  });
}
const Un = /* @__PURE__ */ f("ZodEnum", (e, r) => {
  Wc.init(e, r), P.init(e, r), e.enum = r.entries, e.options = Object.values(r.entries);
  const n = new Set(Object.keys(r.entries));
  e.extract = (t, o) => {
    const u = {};
    for (const c of t)
      if (n.has(c))
        u[c] = r.entries[c];
      else
        throw new Error(`Key ${c} not found in enum`);
    return new Un({
      ...r,
      checks: [],
      ...h(o),
      entries: u
    });
  }, e.exclude = (t, o) => {
    const u = { ...r.entries };
    for (const c of t)
      if (n.has(c))
        delete u[c];
      else
        throw new Error(`Key ${c} not found in enum`);
    return new Un({
      ...r,
      checks: [],
      ...h(o),
      entries: u
    });
  };
});
function Ks(e, r) {
  const n = Array.isArray(e) ? Object.fromEntries(e.map((t) => [t, t])) : e;
  return new Un({
    type: "enum",
    entries: n,
    ...h(r)
  });
}
const Vs = /* @__PURE__ */ f("ZodLiteral", (e, r) => {
  Bc.init(e, r), P.init(e, r), e.values = new Set(r.values), Object.defineProperty(e, "value", {
    get() {
      if (r.values.length > 1)
        throw new Error("This schema contains multiple valid literal values. Use `.values` instead.");
      return r.values[0];
    }
  });
});
function vr(e, r) {
  return new Vs({
    type: "literal",
    values: Array.isArray(e) ? e : [e],
    ...h(r)
  });
}
const Hs = /* @__PURE__ */ f("ZodTransform", (e, r) => {
  Kc.init(e, r), P.init(e, r), e._zod.parse = (n, t) => {
    if (t.direction === "backward")
      throw new bo(e.constructor.name);
    n.addIssue = (u) => {
      if (typeof u == "string")
        n.issues.push(yr(u, n.value, r));
      else {
        const c = u;
        c.fatal && (c.continue = !1), c.code ?? (c.code = "custom"), c.input ?? (c.input = n.value), c.inst ?? (c.inst = e), n.issues.push(yr(c));
      }
    };
    const o = r.transform(n.value, n);
    return o instanceof Promise ? o.then((u) => (n.value = u, n)) : (n.value = o, n);
  };
});
function Gs(e) {
  return new Hs({
    type: "transform",
    transform: e
  });
}
const Uo = /* @__PURE__ */ f("ZodOptional", (e, r) => {
  Vc.init(e, r), P.init(e, r), e.unwrap = () => e._zod.def.innerType;
});
function nr(e) {
  return new Uo({
    type: "optional",
    innerType: e
  });
}
const qs = /* @__PURE__ */ f("ZodNullable", (e, r) => {
  Hc.init(e, r), P.init(e, r), e.unwrap = () => e._zod.def.innerType;
});
function no(e) {
  return new qs({
    type: "nullable",
    innerType: e
  });
}
const Js = /* @__PURE__ */ f("ZodDefault", (e, r) => {
  Gc.init(e, r), P.init(e, r), e.unwrap = () => e._zod.def.innerType, e.removeDefault = e.unwrap;
});
function Ys(e, r) {
  return new Js({
    type: "default",
    innerType: e,
    get defaultValue() {
      return typeof r == "function" ? r() : Oo(r);
    }
  });
}
const Xs = /* @__PURE__ */ f("ZodPrefault", (e, r) => {
  qc.init(e, r), P.init(e, r), e.unwrap = () => e._zod.def.innerType;
});
function Qs(e, r) {
  return new Xs({
    type: "prefault",
    innerType: e,
    get defaultValue() {
      return typeof r == "function" ? r() : Oo(r);
    }
  });
}
const Do = /* @__PURE__ */ f("ZodNonOptional", (e, r) => {
  Jc.init(e, r), P.init(e, r), e.unwrap = () => e._zod.def.innerType;
});
function ef(e, r) {
  return new Do({
    type: "nonoptional",
    innerType: e,
    ...h(r)
  });
}
const rf = /* @__PURE__ */ f("ZodCatch", (e, r) => {
  Yc.init(e, r), P.init(e, r), e.unwrap = () => e._zod.def.innerType, e.removeCatch = e.unwrap;
});
function nf(e, r) {
  return new rf({
    type: "catch",
    innerType: e,
    catchValue: typeof r == "function" ? r : () => r
  });
}
const tf = /* @__PURE__ */ f("ZodPipe", (e, r) => {
  Xc.init(e, r), P.init(e, r), e.in = r.in, e.out = r.out;
});
function to(e, r) {
  return new tf({
    type: "pipe",
    in: e,
    out: r
    // ...util.normalizeParams(params),
  });
}
const of = /* @__PURE__ */ f("ZodReadonly", (e, r) => {
  Qc.init(e, r), P.init(e, r), e.unwrap = () => e._zod.def.innerType;
});
function uf(e) {
  return new of({
    type: "readonly",
    innerType: e
  });
}
const cf = /* @__PURE__ */ f("ZodCustom", (e, r) => {
  ea.init(e, r), P.init(e, r);
});
function af(e, r = {}) {
  return Ka(cf, e, r);
}
function sf(e) {
  return Va(e);
}
Ee({
  app: kr(),
  api: kr()
});
const ff = Ee({
  request: nr(Ee({
    urls: Nn(kr()),
    methods: Nn(Qn([vr("GET"), vr("POST"), vr("PATCH"), vr("PUT"), vr("DELETE")]))
  })),
  log: nr(Ee({})),
  sleep: nr(Ee({}))
});
nr(Ee({
  enabled: Es(),
  requestedScopes: ff
}));
function df(e) {
  return j(e) && !X(e) && !Zr(e) && Symbol.asyncIterator in e;
}
function X(e) {
  return Array.isArray(e);
}
function Lo(e) {
  return typeof e == "bigint";
}
function Sr(e) {
  return typeof e == "boolean";
}
function et(e) {
  return e instanceof globalThis.Date;
}
function mf(e) {
  return typeof e == "function";
}
function lf(e) {
  return j(e) && !X(e) && !Zr(e) && Symbol.iterator in e;
}
function pf(e) {
  return e === null;
}
function be(e) {
  return typeof e == "number";
}
function j(e) {
  return typeof e == "object" && e !== null;
}
function Mo(e) {
  return e instanceof globalThis.RegExp;
}
function T(e) {
  return typeof e == "string";
}
function hf(e) {
  return typeof e == "symbol";
}
function Zr(e) {
  return e instanceof globalThis.Uint8Array;
}
function S(e) {
  return e === void 0;
}
function gf(e) {
  return e.map((r) => en(r));
}
function bf(e) {
  return new Date(e.getTime());
}
function Ff(e) {
  return new Uint8Array(e);
}
function vf(e) {
  return new RegExp(e.source, e.flags);
}
function Of(e) {
  const r = {};
  for (const n of Object.getOwnPropertyNames(e))
    r[n] = en(e[n]);
  for (const n of Object.getOwnPropertySymbols(e))
    r[n] = en(e[n]);
  return r;
}
function en(e) {
  return X(e) ? gf(e) : et(e) ? bf(e) : Zr(e) ? Ff(e) : Mo(e) ? vf(e) : j(e) ? Of(e) : e;
}
function re(e) {
  return en(e);
}
function rt(e, r) {
  return re(r === void 0 ? e : { ...r, ...e });
}
function $f(e) {
  return e !== null && typeof e == "object";
}
function _f(e) {
  return globalThis.Array.isArray(e) && !globalThis.ArrayBuffer.isView(e);
}
function wf(e) {
  return e === void 0;
}
function yf(e) {
  return typeof e == "number";
}
var Dn;
(function(e) {
  e.InstanceMode = "default", e.ExactOptionalPropertyTypes = !1, e.AllowArrayObject = !1, e.AllowNaN = !1, e.AllowNullVoid = !1;
  function r(c, a) {
    return e.ExactOptionalPropertyTypes ? a in c : c[a] !== void 0;
  }
  e.IsExactOptionalProperty = r;
  function n(c) {
    const a = $f(c);
    return e.AllowArrayObject ? a : a && !_f(c);
  }
  e.IsObjectLike = n;
  function t(c) {
    return n(c) && !(c instanceof Date) && !(c instanceof Uint8Array);
  }
  e.IsRecordLike = t;
  function o(c) {
    return e.AllowNaN ? yf(c) : Number.isFinite(c);
  }
  e.IsNumberLike = o;
  function u(c) {
    const a = wf(c);
    return e.AllowNullVoid ? a || c === null : a;
  }
  e.IsVoidLike = u;
})(Dn || (Dn = {}));
function kf(e) {
  return globalThis.Object.freeze(e).map((r) => rn(r));
}
function If(e) {
  const r = {};
  for (const n of Object.getOwnPropertyNames(e))
    r[n] = rn(e[n]);
  for (const n of Object.getOwnPropertySymbols(e))
    r[n] = rn(e[n]);
  return globalThis.Object.freeze(r);
}
function rn(e) {
  return X(e) ? kf(e) : et(e) ? e : Zr(e) ? e : Mo(e) ? e : j(e) ? If(e) : e;
}
function m(e, r) {
  const n = r !== void 0 ? { ...r, ...e } : e;
  switch (Dn.InstanceMode) {
    case "freeze":
      return rn(n);
    case "clone":
      return re(n);
    default:
      return n;
  }
}
class Be extends Error {
  constructor(r) {
    super(r);
  }
}
const ue = Symbol.for("TypeBox.Transform"), jr = Symbol.for("TypeBox.Readonly"), _e = Symbol.for("TypeBox.Optional"), an = Symbol.for("TypeBox.Hint"), g = Symbol.for("TypeBox.Kind");
function nt(e) {
  return j(e) && e[jr] === "Readonly";
}
function Ae(e) {
  return j(e) && e[_e] === "Optional";
}
function Wo(e) {
  return F(e, "Any");
}
function Bo(e) {
  return F(e, "Argument");
}
function ar(e) {
  return F(e, "Array");
}
function sn(e) {
  return F(e, "AsyncIterator");
}
function fn(e) {
  return F(e, "BigInt");
}
function Er(e) {
  return F(e, "Boolean");
}
function sr(e) {
  return F(e, "Computed");
}
function fr(e) {
  return F(e, "Constructor");
}
function Rf(e) {
  return F(e, "Date");
}
function dr(e) {
  return F(e, "Function");
}
function mr(e) {
  return F(e, "Integer");
}
function se(e) {
  return F(e, "Intersect");
}
function dn(e) {
  return F(e, "Iterator");
}
function F(e, r) {
  return j(e) && g in e && e[g] === r;
}
function Ko(e) {
  return Sr(e) || be(e) || T(e);
}
function Ke(e) {
  return F(e, "Literal");
}
function Ve(e) {
  return F(e, "MappedKey");
}
function oe(e) {
  return F(e, "MappedResult");
}
function Cr(e) {
  return F(e, "Never");
}
function xf(e) {
  return F(e, "Not");
}
function tt(e) {
  return F(e, "Null");
}
function lr(e) {
  return F(e, "Number");
}
function he(e) {
  return F(e, "Object");
}
function mn(e) {
  return F(e, "Promise");
}
function ln(e) {
  return F(e, "Record");
}
function q(e) {
  return F(e, "Ref");
}
function Vo(e) {
  return F(e, "RegExp");
}
function Nr(e) {
  return F(e, "String");
}
function ot(e) {
  return F(e, "Symbol");
}
function He(e) {
  return F(e, "TemplateLiteral");
}
function Af(e) {
  return F(e, "This");
}
function pn(e) {
  return j(e) && ue in e;
}
function Ge(e) {
  return F(e, "Tuple");
}
function it(e) {
  return F(e, "Undefined");
}
function L(e) {
  return F(e, "Union");
}
function zf(e) {
  return F(e, "Uint8Array");
}
function Tf(e) {
  return F(e, "Unknown");
}
function Pf(e) {
  return F(e, "Unsafe");
}
function Sf(e) {
  return F(e, "Void");
}
function Zf(e) {
  return j(e) && g in e && T(e[g]);
}
function Re(e) {
  return Wo(e) || Bo(e) || ar(e) || Er(e) || fn(e) || sn(e) || sr(e) || fr(e) || Rf(e) || dr(e) || mr(e) || se(e) || dn(e) || Ke(e) || Ve(e) || oe(e) || Cr(e) || xf(e) || tt(e) || lr(e) || he(e) || mn(e) || ln(e) || q(e) || Vo(e) || Nr(e) || ot(e) || He(e) || Af(e) || Ge(e) || it(e) || L(e) || zf(e) || Tf(e) || Pf(e) || Sf(e) || Zf(e);
}
const jf = [
  "Argument",
  "Any",
  "Array",
  "AsyncIterator",
  "BigInt",
  "Boolean",
  "Computed",
  "Constructor",
  "Date",
  "Enum",
  "Function",
  "Integer",
  "Intersect",
  "Iterator",
  "Literal",
  "MappedKey",
  "MappedResult",
  "Not",
  "Null",
  "Number",
  "Object",
  "Promise",
  "Record",
  "Ref",
  "RegExp",
  "String",
  "Symbol",
  "TemplateLiteral",
  "This",
  "Tuple",
  "Undefined",
  "Union",
  "Uint8Array",
  "Unknown",
  "Void"
];
function Ho(e) {
  try {
    return new RegExp(e), !0;
  } catch {
    return !1;
  }
}
function ut(e) {
  if (!T(e))
    return !1;
  for (let r = 0; r < e.length; r++) {
    const n = e.charCodeAt(r);
    if (n >= 7 && n <= 13 || n === 27 || n === 127)
      return !1;
  }
  return !0;
}
function Go(e) {
  return ct(e) || W(e);
}
function Or(e) {
  return S(e) || Lo(e);
}
function w(e) {
  return S(e) || be(e);
}
function ct(e) {
  return S(e) || Sr(e);
}
function _(e) {
  return S(e) || T(e);
}
function Ef(e) {
  return S(e) || T(e) && ut(e) && Ho(e);
}
function Cf(e) {
  return S(e) || T(e) && ut(e);
}
function qo(e) {
  return S(e) || W(e);
}
function nn(e) {
  return j(e) && e[_e] === "Optional";
}
function de(e) {
  return v(e, "Any") && _(e.$id);
}
function Nf(e) {
  return v(e, "Argument") && be(e.index);
}
function qe(e) {
  return v(e, "Array") && e.type === "array" && _(e.$id) && W(e.items) && w(e.minItems) && w(e.maxItems) && ct(e.uniqueItems) && qo(e.contains) && w(e.minContains) && w(e.maxContains);
}
function at(e) {
  return v(e, "AsyncIterator") && e.type === "AsyncIterator" && _(e.$id) && W(e.items);
}
function hn(e) {
  return v(e, "BigInt") && e.type === "bigint" && _(e.$id) && Or(e.exclusiveMaximum) && Or(e.exclusiveMinimum) && Or(e.maximum) && Or(e.minimum) && Or(e.multipleOf);
}
function Je(e) {
  return v(e, "Boolean") && e.type === "boolean" && _(e.$id);
}
function Uf(e) {
  return v(e, "Computed") && T(e.target) && X(e.parameters) && e.parameters.every((r) => W(r));
}
function gn(e) {
  return v(e, "Constructor") && e.type === "Constructor" && _(e.$id) && X(e.parameters) && e.parameters.every((r) => W(r)) && W(e.returns);
}
function bn(e) {
  return v(e, "Date") && e.type === "Date" && _(e.$id) && w(e.exclusiveMaximumTimestamp) && w(e.exclusiveMinimumTimestamp) && w(e.maximumTimestamp) && w(e.minimumTimestamp) && w(e.multipleOfTimestamp);
}
function Fn(e) {
  return v(e, "Function") && e.type === "Function" && _(e.$id) && X(e.parameters) && e.parameters.every((r) => W(r)) && W(e.returns);
}
function we(e) {
  return v(e, "Integer") && e.type === "integer" && _(e.$id) && w(e.exclusiveMaximum) && w(e.exclusiveMinimum) && w(e.maximum) && w(e.minimum) && w(e.multipleOf);
}
function Jo(e) {
  return j(e) && Object.entries(e).every(([r, n]) => ut(r) && W(n));
}
function Ye(e) {
  return v(e, "Intersect") && !(T(e.type) && e.type !== "object") && X(e.allOf) && e.allOf.every((r) => W(r) && !Kf(r)) && _(e.type) && (ct(e.unevaluatedProperties) || qo(e.unevaluatedProperties)) && _(e.$id);
}
function st(e) {
  return v(e, "Iterator") && e.type === "Iterator" && _(e.$id) && W(e.items);
}
function v(e, r) {
  return j(e) && g in e && e[g] === r;
}
function Yo(e) {
  return ze(e) && T(e.const);
}
function Xo(e) {
  return ze(e) && be(e.const);
}
function Qo(e) {
  return ze(e) && Sr(e.const);
}
function ze(e) {
  return v(e, "Literal") && _(e.$id) && Df(e.const);
}
function Df(e) {
  return Sr(e) || be(e) || T(e);
}
function Lf(e) {
  return v(e, "MappedKey") && X(e.keys) && e.keys.every((r) => be(r) || T(r));
}
function Mf(e) {
  return v(e, "MappedResult") && Jo(e.properties);
}
function Te(e) {
  return v(e, "Never") && j(e.not) && Object.getOwnPropertyNames(e.not).length === 0;
}
function or(e) {
  return v(e, "Not") && W(e.not);
}
function ft(e) {
  return v(e, "Null") && e.type === "null" && _(e.$id);
}
function G(e) {
  return v(e, "Number") && e.type === "number" && _(e.$id) && w(e.exclusiveMaximum) && w(e.exclusiveMinimum) && w(e.maximum) && w(e.minimum) && w(e.multipleOf);
}
function A(e) {
  return v(e, "Object") && e.type === "object" && _(e.$id) && Jo(e.properties) && Go(e.additionalProperties) && w(e.minProperties) && w(e.maxProperties);
}
function dt(e) {
  return v(e, "Promise") && e.type === "Promise" && _(e.$id) && W(e.item);
}
function M(e) {
  return v(e, "Record") && e.type === "object" && _(e.$id) && Go(e.additionalProperties) && j(e.patternProperties) && ((r) => {
    const n = Object.getOwnPropertyNames(r.patternProperties);
    return n.length === 1 && Ho(n[0]) && j(r.patternProperties) && W(r.patternProperties[n[0]]);
  })(e);
}
function Wf(e) {
  return v(e, "Ref") && _(e.$id) && T(e.$ref);
}
function Ir(e) {
  return v(e, "RegExp") && _(e.$id) && T(e.source) && T(e.flags) && w(e.maxLength) && w(e.minLength);
}
function me(e) {
  return v(e, "String") && e.type === "string" && _(e.$id) && w(e.minLength) && w(e.maxLength) && Ef(e.pattern) && Cf(e.format);
}
function Rr(e) {
  return v(e, "Symbol") && e.type === "symbol" && _(e.$id);
}
function xr(e) {
  return v(e, "TemplateLiteral") && e.type === "string" && T(e.pattern) && e.pattern[0] === "^" && e.pattern[e.pattern.length - 1] === "$";
}
function Bf(e) {
  return v(e, "This") && _(e.$id) && T(e.$ref);
}
function Kf(e) {
  return j(e) && ue in e;
}
function vn(e) {
  return v(e, "Tuple") && e.type === "array" && _(e.$id) && be(e.minItems) && be(e.maxItems) && e.minItems === e.maxItems && // empty
  (S(e.items) && S(e.additionalItems) && e.minItems === 0 || X(e.items) && e.items.every((r) => W(r)));
}
function De(e) {
  return v(e, "Undefined") && e.type === "undefined" && _(e.$id);
}
function $e(e) {
  return v(e, "Union") && _(e.$id) && j(e) && X(e.anyOf) && e.anyOf.every((r) => W(r));
}
function Ur(e) {
  return v(e, "Uint8Array") && e.type === "Uint8Array" && _(e.$id) && w(e.minByteLength) && w(e.maxByteLength);
}
function le(e) {
  return v(e, "Unknown") && _(e.$id);
}
function Vf(e) {
  return v(e, "Unsafe");
}
function On(e) {
  return v(e, "Void") && e.type === "void" && _(e.$id);
}
function Hf(e) {
  return j(e) && g in e && T(e[g]) && !jf.includes(e[g]);
}
function W(e) {
  return j(e) && (de(e) || Nf(e) || qe(e) || Je(e) || hn(e) || at(e) || Uf(e) || gn(e) || bn(e) || Fn(e) || we(e) || Ye(e) || st(e) || ze(e) || Lf(e) || Mf(e) || Te(e) || or(e) || ft(e) || G(e) || A(e) || dt(e) || M(e) || Wf(e) || Ir(e) || me(e) || Rr(e) || xr(e) || Bf(e) || vn(e) || De(e) || $e(e) || Ur(e) || le(e) || Vf(e) || On(e) || Hf(e));
}
const Gf = "(true|false)", Jr = "(0|[1-9][0-9]*)", ei = "(.*)", qf = "(?!.*)", ir = `^${Jr}$`, ur = `^${ei}$`, Jf = `^${qf}$`;
function Yf(e, r) {
  return e.includes(r);
}
function Xf(e) {
  return [...new Set(e)];
}
function Qf(e, r) {
  return e.filter((n) => r.includes(n));
}
function ed(e, r) {
  return e.reduce((n, t) => Qf(n, t), r);
}
function rd(e) {
  return e.length === 1 ? e[0] : e.length > 1 ? ed(e.slice(1), e[0]) : [];
}
function nd(e) {
  const r = [];
  for (const n of e)
    r.push(...n);
  return r;
}
function Ar(e) {
  return m({ [g]: "Any" }, e);
}
function mt(e, r) {
  return m({ [g]: "Array", type: "array", items: e }, r);
}
function td(e) {
  return m({ [g]: "Argument", index: e });
}
function lt(e, r) {
  return m({ [g]: "AsyncIterator", type: "AsyncIterator", items: e }, r);
}
function U(e, r, n) {
  return m({ [g]: "Computed", target: e, parameters: r }, n);
}
function od(e, r) {
  const { [r]: n, ...t } = e;
  return t;
}
function ne(e, r) {
  return r.reduce((n, t) => od(n, t), e);
}
function E(e) {
  return m({ [g]: "Never", not: {} }, e);
}
function B(e) {
  return m({
    [g]: "MappedResult",
    properties: e
  });
}
function pt(e, r, n) {
  return m({ [g]: "Constructor", type: "Constructor", parameters: e, returns: r }, n);
}
function Dr(e, r, n) {
  return m({ [g]: "Function", type: "Function", parameters: e, returns: r }, n);
}
function Ln(e, r) {
  return m({ [g]: "Union", anyOf: e }, r);
}
function id(e) {
  return e.some((r) => Ae(r));
}
function oo(e) {
  return e.map((r) => Ae(r) ? ud(r) : r);
}
function ud(e) {
  return ne(e, [_e]);
}
function cd(e, r) {
  return id(e) ? Ze(Ln(oo(e), r)) : Ln(oo(e), r);
}
function pr(e, r) {
  return e.length === 1 ? m(e[0], r) : e.length === 0 ? E(r) : cd(e, r);
}
function K(e, r) {
  return e.length === 0 ? E(r) : e.length === 1 ? m(e[0], r) : Ln(e, r);
}
class io extends Be {
}
function ad(e) {
  return e.replace(/\\\$/g, "$").replace(/\\\*/g, "*").replace(/\\\^/g, "^").replace(/\\\|/g, "|").replace(/\\\(/g, "(").replace(/\\\)/g, ")");
}
function ht(e, r, n) {
  return e[r] === n && e.charCodeAt(r - 1) !== 92;
}
function Oe(e, r) {
  return ht(e, r, "(");
}
function zr(e, r) {
  return ht(e, r, ")");
}
function ri(e, r) {
  return ht(e, r, "|");
}
function sd(e) {
  if (!(Oe(e, 0) && zr(e, e.length - 1)))
    return !1;
  let r = 0;
  for (let n = 0; n < e.length; n++)
    if (Oe(e, n) && (r += 1), zr(e, n) && (r -= 1), r === 0 && n !== e.length - 1)
      return !1;
  return !0;
}
function fd(e) {
  return e.slice(1, e.length - 1);
}
function dd(e) {
  let r = 0;
  for (let n = 0; n < e.length; n++)
    if (Oe(e, n) && (r += 1), zr(e, n) && (r -= 1), ri(e, n) && r === 0)
      return !0;
  return !1;
}
function md(e) {
  for (let r = 0; r < e.length; r++)
    if (Oe(e, r))
      return !0;
  return !1;
}
function ld(e) {
  let [r, n] = [0, 0];
  const t = [];
  for (let u = 0; u < e.length; u++)
    if (Oe(e, u) && (r += 1), zr(e, u) && (r -= 1), ri(e, u) && r === 0) {
      const c = e.slice(n, u);
      c.length > 0 && t.push(cr(c)), n = u + 1;
    }
  const o = e.slice(n);
  return o.length > 0 && t.push(cr(o)), t.length === 0 ? { type: "const", const: "" } : t.length === 1 ? t[0] : { type: "or", expr: t };
}
function pd(e) {
  function r(o, u) {
    if (!Oe(o, u))
      throw new io("TemplateLiteralParser: Index must point to open parens");
    let c = 0;
    for (let a = u; a < o.length; a++)
      if (Oe(o, a) && (c += 1), zr(o, a) && (c -= 1), c === 0)
        return [u, a];
    throw new io("TemplateLiteralParser: Unclosed group parens in expression");
  }
  function n(o, u) {
    for (let c = u; c < o.length; c++)
      if (Oe(o, c))
        return [u, c];
    return [u, o.length];
  }
  const t = [];
  for (let o = 0; o < e.length; o++)
    if (Oe(e, o)) {
      const [u, c] = r(e, o), a = e.slice(u, c + 1);
      t.push(cr(a)), o = c;
    } else {
      const [u, c] = n(e, o), a = e.slice(u, c);
      a.length > 0 && t.push(cr(a)), o = c - 1;
    }
  return t.length === 0 ? { type: "const", const: "" } : t.length === 1 ? t[0] : { type: "and", expr: t };
}
function cr(e) {
  return sd(e) ? cr(fd(e)) : dd(e) ? ld(e) : md(e) ? pd(e) : { type: "const", const: ad(e) };
}
function gt(e) {
  return cr(e.slice(1, e.length - 1));
}
class hd extends Be {
}
function gd(e) {
  return e.type === "or" && e.expr.length === 2 && e.expr[0].type === "const" && e.expr[0].const === "0" && e.expr[1].type === "const" && e.expr[1].const === "[1-9][0-9]*";
}
function bd(e) {
  return e.type === "or" && e.expr.length === 2 && e.expr[0].type === "const" && e.expr[0].const === "true" && e.expr[1].type === "const" && e.expr[1].const === "false";
}
function Fd(e) {
  return e.type === "const" && e.const === ".*";
}
function Tr(e) {
  return gd(e) || Fd(e) ? !1 : bd(e) ? !0 : e.type === "and" ? e.expr.every((r) => Tr(r)) : e.type === "or" ? e.expr.every((r) => Tr(r)) : e.type === "const" ? !0 : (() => {
    throw new hd("Unknown expression type");
  })();
}
function vd(e) {
  const r = gt(e.pattern);
  return Tr(r);
}
class Od extends Be {
}
function* ni(e) {
  if (e.length === 1)
    return yield* e[0];
  for (const r of e[0])
    for (const n of ni(e.slice(1)))
      yield `${r}${n}`;
}
function* $d(e) {
  return yield* ni(e.expr.map((r) => [...$n(r)]));
}
function* _d(e) {
  for (const r of e.expr)
    yield* $n(r);
}
function* wd(e) {
  return yield e.const;
}
function* $n(e) {
  return e.type === "and" ? yield* $d(e) : e.type === "or" ? yield* _d(e) : e.type === "const" ? yield* wd(e) : (() => {
    throw new Od("Unknown expression");
  })();
}
function ti(e) {
  const r = gt(e.pattern);
  return Tr(r) ? [...$n(r)] : [];
}
function Z(e, r) {
  return m({
    [g]: "Literal",
    const: e,
    type: typeof e
  }, r);
}
function oi(e) {
  return m({ [g]: "Boolean", type: "boolean" }, e);
}
function bt(e) {
  return m({ [g]: "BigInt", type: "bigint" }, e);
}
function Xe(e) {
  return m({ [g]: "Number", type: "number" }, e);
}
function Le(e) {
  return m({ [g]: "String", type: "string" }, e);
}
function* yd(e) {
  const r = e.trim().replace(/"|'/g, "");
  return r === "boolean" ? yield oi() : r === "number" ? yield Xe() : r === "bigint" ? yield bt() : r === "string" ? yield Le() : yield (() => {
    const n = r.split("|").map((t) => Z(t.trim()));
    return n.length === 0 ? E() : n.length === 1 ? n[0] : pr(n);
  })();
}
function* kd(e) {
  if (e[1] !== "{") {
    const r = Z("$"), n = Mn(e.slice(1));
    return yield* [r, ...n];
  }
  for (let r = 2; r < e.length; r++)
    if (e[r] === "}") {
      const n = yd(e.slice(2, r)), t = Mn(e.slice(r + 1));
      return yield* [...n, ...t];
    }
  yield Z(e);
}
function* Mn(e) {
  for (let r = 0; r < e.length; r++)
    if (e[r] === "$") {
      const n = Z(e.slice(0, r)), t = kd(e.slice(r));
      return yield* [n, ...t];
    }
  yield Z(e);
}
function Id(e) {
  return [...Mn(e)];
}
class Rd extends Be {
}
function xd(e) {
  return e.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}
function ii(e, r) {
  return He(e) ? e.pattern.slice(1, e.pattern.length - 1) : L(e) ? `(${e.anyOf.map((n) => ii(n, r)).join("|")})` : lr(e) ? `${r}${Jr}` : mr(e) ? `${r}${Jr}` : fn(e) ? `${r}${Jr}` : Nr(e) ? `${r}${ei}` : Ke(e) ? `${r}${xd(e.const.toString())}` : Er(e) ? `${r}${Gf}` : (() => {
    throw new Rd(`Unexpected Kind '${e[g]}'`);
  })();
}
function uo(e) {
  return `^${e.map((r) => ii(r, "")).join("")}$`;
}
function tn(e) {
  const n = ti(e).map((t) => Z(t));
  return pr(n);
}
function ui(e, r) {
  const n = T(e) ? uo(Id(e)) : uo(e);
  return m({ [g]: "TemplateLiteral", type: "string", pattern: n }, r);
}
function Ad(e) {
  return ti(e).map((n) => n.toString());
}
function zd(e) {
  const r = [];
  for (const n of e)
    r.push(...Pe(n));
  return r;
}
function Td(e) {
  return [e.toString()];
}
function Pe(e) {
  return [...new Set(He(e) ? Ad(e) : L(e) ? zd(e.anyOf) : Ke(e) ? Td(e.const) : lr(e) ? ["[number]"] : mr(e) ? ["[number]"] : [])];
}
function Pd(e, r, n) {
  const t = {};
  for (const o of Object.getOwnPropertyNames(r))
    t[o] = _n(e, Pe(r[o]), n);
  return t;
}
function Sd(e, r, n) {
  return Pd(e, r.properties, n);
}
function Zd(e, r, n) {
  const t = Sd(e, r, n);
  return B(t);
}
function ci(e, r) {
  return e.map((n) => ai(n, r));
}
function jd(e) {
  return e.filter((r) => !Cr(r));
}
function Ed(e, r) {
  return mi(jd(ci(e, r)));
}
function Cd(e) {
  return e.some((r) => Cr(r)) ? [] : e;
}
function Nd(e, r) {
  return pr(Cd(ci(e, r)));
}
function Ud(e, r) {
  return r in e ? e[r] : r === "[number]" ? pr(e) : E();
}
function Dd(e, r) {
  return r === "[number]" ? e : E();
}
function Ld(e, r) {
  return r in e ? e[r] : E();
}
function ai(e, r) {
  return se(e) ? Ed(e.allOf, r) : L(e) ? Nd(e.anyOf, r) : Ge(e) ? Ud(e.items ?? [], r) : ar(e) ? Dd(e.items, r) : he(e) ? Ld(e.properties, r) : E();
}
function si(e, r) {
  return r.map((n) => ai(e, n));
}
function co(e, r) {
  return pr(si(e, r));
}
function _n(e, r, n) {
  if (q(e) || q(r)) {
    const t = "Index types using Ref parameters require both Type and Key to be of TSchema";
    if (!Re(e) || !Re(r))
      throw new Be(t);
    return U("Index", [e, r]);
  }
  return oe(r) ? Zd(e, r, n) : Ve(r) ? Kd(e, r, n) : m(Re(r) ? co(e, Pe(r)) : co(e, r), n);
}
function Md(e, r, n) {
  return { [r]: _n(e, [r], re(n)) };
}
function Wd(e, r, n) {
  return r.reduce((t, o) => ({ ...t, ...Md(e, o, n) }), {});
}
function Bd(e, r, n) {
  return Wd(e, r.keys, n);
}
function Kd(e, r, n) {
  const t = Bd(e, r, n);
  return B(t);
}
function Ft(e, r) {
  return m({ [g]: "Iterator", type: "Iterator", items: e }, r);
}
function Vd(e) {
  const r = [];
  for (let n in e)
    Ae(e[n]) || r.push(n);
  return r;
}
function Hd(e, r) {
  const n = Vd(e), t = n.length > 0 ? { [g]: "Object", type: "object", properties: e, required: n } : { [g]: "Object", type: "object", properties: e };
  return m(t, r);
}
var D = Hd;
function fi(e, r) {
  return m({ [g]: "Promise", type: "Promise", item: e }, r);
}
function Gd(e) {
  return m(ne(e, [jr]));
}
function qd(e) {
  return m({ ...e, [jr]: "Readonly" });
}
function Jd(e, r) {
  return r === !1 ? Gd(e) : qd(e);
}
function Se(e, r) {
  const n = r ?? !0;
  return oe(e) ? Qd(e, n) : Jd(e, n);
}
function Yd(e, r) {
  const n = {};
  for (const t of globalThis.Object.getOwnPropertyNames(e))
    n[t] = Se(e[t], r);
  return n;
}
function Xd(e, r) {
  return Yd(e.properties, r);
}
function Qd(e, r) {
  const n = Xd(e, r);
  return B(n);
}
function hr(e, r) {
  return m(e.length > 0 ? { [g]: "Tuple", type: "array", items: e, additionalItems: !1, minItems: e.length, maxItems: e.length } : { [g]: "Tuple", type: "array", minItems: e.length, maxItems: e.length }, r);
}
function di(e, r) {
  return e in r ? ie(e, r[e]) : B(r);
}
function em(e) {
  return { [e]: Z(e) };
}
function rm(e) {
  const r = {};
  for (const n of e)
    r[n] = Z(n);
  return r;
}
function nm(e, r) {
  return Yf(r, e) ? em(e) : rm(r);
}
function tm(e, r) {
  const n = nm(e, r);
  return di(e, n);
}
function $r(e, r) {
  return r.map((n) => ie(e, n));
}
function om(e, r) {
  const n = {};
  for (const t of globalThis.Object.getOwnPropertyNames(r))
    n[t] = ie(e, r[t]);
  return n;
}
function ie(e, r) {
  const n = { ...r };
  return (
    // unevaluated modifier types
    Ae(r) ? Ze(ie(e, ne(r, [_e]))) : nt(r) ? Se(ie(e, ne(r, [jr]))) : (
      // unevaluated mapped types
      oe(r) ? di(e, r.properties) : Ve(r) ? tm(e, r.keys) : (
        // unevaluated types
        fr(r) ? pt($r(e, r.parameters), ie(e, r.returns), n) : dr(r) ? Dr($r(e, r.parameters), ie(e, r.returns), n) : sn(r) ? lt(ie(e, r.items), n) : dn(r) ? Ft(ie(e, r.items), n) : se(r) ? je($r(e, r.allOf), n) : L(r) ? K($r(e, r.anyOf), n) : Ge(r) ? hr($r(e, r.items ?? []), n) : he(r) ? D(om(e, r.properties), n) : ar(r) ? mt(ie(e, r.items), n) : mn(r) ? fi(ie(e, r.item), n) : r
      )
    )
  );
}
function im(e, r) {
  const n = {};
  for (const t of e)
    n[t] = ie(t, r);
  return n;
}
function um(e, r, n) {
  const t = Re(e) ? Pe(e) : e, o = r({ [g]: "MappedKey", keys: t }), u = im(t, o);
  return D(u, n);
}
function cm(e) {
  return m(ne(e, [_e]));
}
function am(e) {
  return m({ ...e, [_e]: "Optional" });
}
function sm(e, r) {
  return r === !1 ? cm(e) : am(e);
}
function Ze(e, r) {
  const n = r ?? !0;
  return oe(e) ? mm(e, n) : sm(e, n);
}
function fm(e, r) {
  const n = {};
  for (const t of globalThis.Object.getOwnPropertyNames(e))
    n[t] = Ze(e[t], r);
  return n;
}
function dm(e, r) {
  return fm(e.properties, r);
}
function mm(e, r) {
  const n = dm(e, r);
  return B(n);
}
function Wn(e, r = {}) {
  const n = e.every((o) => he(o)), t = Re(r.unevaluatedProperties) ? { unevaluatedProperties: r.unevaluatedProperties } : {};
  return m(r.unevaluatedProperties === !1 || Re(r.unevaluatedProperties) || n ? { ...t, [g]: "Intersect", type: "object", allOf: e } : { ...t, [g]: "Intersect", allOf: e }, r);
}
function lm(e) {
  return e.every((r) => Ae(r));
}
function pm(e) {
  return ne(e, [_e]);
}
function ao(e) {
  return e.map((r) => Ae(r) ? pm(r) : r);
}
function hm(e, r) {
  return lm(e) ? Ze(Wn(ao(e), r)) : Wn(ao(e), r);
}
function mi(e, r = {}) {
  if (e.length === 1)
    return m(e[0], r);
  if (e.length === 0)
    return E(r);
  if (e.some((n) => pn(n)))
    throw new Error("Cannot intersect transform types");
  return hm(e, r);
}
function je(e, r) {
  if (e.length === 1)
    return m(e[0], r);
  if (e.length === 0)
    return E(r);
  if (e.some((n) => pn(n)))
    throw new Error("Cannot intersect transform types");
  return Wn(e, r);
}
function Lr(...e) {
  const [r, n] = typeof e[0] == "string" ? [e[0], e[1]] : [e[0].$id, e[1]];
  if (typeof r != "string")
    throw new Be("Ref: $ref must be a string");
  return m({ [g]: "Ref", $ref: r }, n);
}
function gm(e, r) {
  return U("Awaited", [U(e, r)]);
}
function bm(e) {
  return U("Awaited", [Lr(e)]);
}
function Fm(e) {
  return je(li(e));
}
function vm(e) {
  return K(li(e));
}
function Om(e) {
  return wn(e);
}
function li(e) {
  return e.map((r) => wn(r));
}
function wn(e, r) {
  return m(sr(e) ? gm(e.target, e.parameters) : se(e) ? Fm(e.allOf) : L(e) ? vm(e.anyOf) : mn(e) ? Om(e.item) : q(e) ? bm(e.$ref) : e, r);
}
function pi(e) {
  const r = [];
  for (const n of e)
    r.push(vt(n));
  return r;
}
function $m(e) {
  const r = pi(e);
  return nd(r);
}
function _m(e) {
  const r = pi(e);
  return rd(r);
}
function wm(e) {
  return e.map((r, n) => n.toString());
}
function ym(e) {
  return ["[number]"];
}
function km(e) {
  return globalThis.Object.getOwnPropertyNames(e);
}
function Im(e) {
  return [];
}
function vt(e) {
  return se(e) ? $m(e.allOf) : L(e) ? _m(e.anyOf) : Ge(e) ? wm(e.items ?? []) : ar(e) ? ym(e.items) : he(e) ? km(e.properties) : ln(e) ? Im(e.patternProperties) : [];
}
function Rm(e, r) {
  return U("KeyOf", [U(e, r)]);
}
function xm(e) {
  return U("KeyOf", [Lr(e)]);
}
function Am(e, r) {
  const n = vt(e), t = zm(n), o = pr(t);
  return m(o, r);
}
function zm(e) {
  return e.map((r) => r === "[number]" ? Xe() : Z(r));
}
function Ot(e, r) {
  return sr(e) ? Rm(e.target, e.parameters) : q(e) ? xm(e.$ref) : oe(e) ? Sm(e, r) : Am(e, r);
}
function Tm(e, r) {
  const n = {};
  for (const t of globalThis.Object.getOwnPropertyNames(e))
    n[t] = Ot(e[t], re(r));
  return n;
}
function Pm(e, r) {
  return Tm(e.properties, r);
}
function Sm(e, r) {
  const n = Pm(e, r);
  return B(n);
}
function Zm(e) {
  const r = [];
  for (const n of e)
    r.push(...vt(n));
  return Xf(r);
}
function jm(e) {
  return e.filter((r) => !Cr(r));
}
function Em(e, r) {
  const n = [];
  for (const t of e)
    n.push(...si(t, [r]));
  return jm(n);
}
function Cm(e, r) {
  const n = {};
  for (const t of r)
    n[t] = mi(Em(e, t));
  return n;
}
function Nm(e, r) {
  const n = Zm(e), t = Cm(e, n);
  return D(t, r);
}
function hi(e) {
  return m({ [g]: "Date", type: "Date" }, e);
}
function gi(e) {
  return m({ [g]: "Null", type: "null" }, e);
}
function bi(e) {
  return m({ [g]: "Symbol", type: "symbol" }, e);
}
function Fi(e) {
  return m({ [g]: "Undefined", type: "undefined" }, e);
}
function vi(e) {
  return m({ [g]: "Uint8Array", type: "Uint8Array" }, e);
}
function yn(e) {
  return m({ [g]: "Unknown" }, e);
}
function Um(e) {
  return e.map((r) => $t(r, !1));
}
function Dm(e) {
  const r = {};
  for (const n of globalThis.Object.getOwnPropertyNames(e))
    r[n] = Se($t(e[n], !1));
  return r;
}
function qr(e, r) {
  return r === !0 ? e : Se(e);
}
function $t(e, r) {
  return df(e) || lf(e) ? qr(Ar(), r) : X(e) ? Se(hr(Um(e))) : Zr(e) ? vi() : et(e) ? hi() : j(e) ? qr(D(Dm(e)), r) : mf(e) ? qr(Dr([], yn()), r) : S(e) ? Fi() : pf(e) ? gi() : hf(e) ? bi() : Lo(e) ? bt() : be(e) || Sr(e) || T(e) ? Z(e) : D({});
}
function Lm(e, r) {
  return m($t(e, !0), r);
}
function Mm(e, r) {
  return fr(e) ? hr(e.parameters, r) : E(r);
}
function Wm(e, r) {
  if (S(e))
    throw new Error("Enum undefined or empty");
  const n = globalThis.Object.getOwnPropertyNames(e).filter((u) => isNaN(u)).map((u) => e[u]), o = [...new Set(n)].map((u) => Z(u));
  return K(o, { ...r, [an]: "Enum" });
}
class Bm extends Be {
}
var s;
(function(e) {
  e[e.Union = 0] = "Union", e[e.True = 1] = "True", e[e.False = 2] = "False";
})(s || (s = {}));
function ce(e) {
  return e === s.False ? e : s.True;
}
function gr(e) {
  throw new Bm(e);
}
function C(e) {
  return Te(e) || Ye(e) || $e(e) || le(e) || de(e);
}
function N(e, r) {
  return Te(r) ? _i() : Ye(r) ? kn(e, r) : $e(r) ? wt(e, r) : le(r) ? Ii() : de(r) ? _t() : gr("StructuralRight");
}
function _t(e, r) {
  return s.True;
}
function Km(e, r) {
  return Ye(r) ? kn(e, r) : $e(r) && r.anyOf.some((n) => de(n) || le(n)) ? s.True : $e(r) ? s.Union : le(r) || de(r) ? s.True : s.Union;
}
function Vm(e, r) {
  return le(e) ? s.False : de(e) ? s.Union : Te(e) ? s.True : s.False;
}
function Hm(e, r) {
  return A(r) && In(r) ? s.True : C(r) ? N(e, r) : qe(r) ? ce($(e.items, r.items)) : s.False;
}
function Gm(e, r) {
  return C(r) ? N(e, r) : at(r) ? ce($(e.items, r.items)) : s.False;
}
function qm(e, r) {
  return C(r) ? N(e, r) : A(r) ? V(e, r) : M(r) ? fe(e, r) : hn(r) ? s.True : s.False;
}
function Oi(e, r) {
  return Qo(e) || Je(e) ? s.True : s.False;
}
function Jm(e, r) {
  return C(r) ? N(e, r) : A(r) ? V(e, r) : M(r) ? fe(e, r) : Je(r) ? s.True : s.False;
}
function Ym(e, r) {
  return C(r) ? N(e, r) : A(r) ? V(e, r) : gn(r) ? e.parameters.length > r.parameters.length ? s.False : e.parameters.every((n, t) => ce($(r.parameters[t], n)) === s.True) ? ce($(e.returns, r.returns)) : s.False : s.False;
}
function Xm(e, r) {
  return C(r) ? N(e, r) : A(r) ? V(e, r) : M(r) ? fe(e, r) : bn(r) ? s.True : s.False;
}
function Qm(e, r) {
  return C(r) ? N(e, r) : A(r) ? V(e, r) : Fn(r) ? e.parameters.length > r.parameters.length ? s.False : e.parameters.every((n, t) => ce($(r.parameters[t], n)) === s.True) ? ce($(e.returns, r.returns)) : s.False : s.False;
}
function $i(e, r) {
  return ze(e) && be(e.const) || G(e) || we(e) ? s.True : s.False;
}
function el(e, r) {
  return we(r) || G(r) ? s.True : C(r) ? N(e, r) : A(r) ? V(e, r) : M(r) ? fe(e, r) : s.False;
}
function kn(e, r) {
  return r.allOf.every((n) => $(e, n) === s.True) ? s.True : s.False;
}
function rl(e, r) {
  return e.allOf.some((n) => $(n, r) === s.True) ? s.True : s.False;
}
function nl(e, r) {
  return C(r) ? N(e, r) : st(r) ? ce($(e.items, r.items)) : s.False;
}
function tl(e, r) {
  return ze(r) && r.const === e.const ? s.True : C(r) ? N(e, r) : A(r) ? V(e, r) : M(r) ? fe(e, r) : me(r) ? ki(e) : G(r) ? wi(e) : we(r) ? $i(e) : Je(r) ? Oi(e) : s.False;
}
function _i(e, r) {
  return s.False;
}
function ol(e, r) {
  return s.True;
}
function so(e) {
  let [r, n] = [e, 0];
  for (; or(r); )
    r = r.not, n += 1;
  return n % 2 === 0 ? r : yn();
}
function il(e, r) {
  return or(e) ? $(so(e), r) : or(r) ? $(e, so(r)) : gr("Invalid fallthrough for Not");
}
function ul(e, r) {
  return C(r) ? N(e, r) : A(r) ? V(e, r) : M(r) ? fe(e, r) : ft(r) ? s.True : s.False;
}
function wi(e, r) {
  return Xo(e) || G(e) || we(e) ? s.True : s.False;
}
function cl(e, r) {
  return C(r) ? N(e, r) : A(r) ? V(e, r) : M(r) ? fe(e, r) : we(r) || G(r) ? s.True : s.False;
}
function J(e, r) {
  return Object.getOwnPropertyNames(e.properties).length === r;
}
function fo(e) {
  return In(e);
}
function mo(e) {
  return J(e, 0) || J(e, 1) && "description" in e.properties && $e(e.properties.description) && e.properties.description.anyOf.length === 2 && (me(e.properties.description.anyOf[0]) && De(e.properties.description.anyOf[1]) || me(e.properties.description.anyOf[1]) && De(e.properties.description.anyOf[0]));
}
function Zn(e) {
  return J(e, 0);
}
function lo(e) {
  return J(e, 0);
}
function al(e) {
  return J(e, 0);
}
function sl(e) {
  return J(e, 0);
}
function fl(e) {
  return In(e);
}
function dl(e) {
  const r = Xe();
  return J(e, 0) || J(e, 1) && "length" in e.properties && ce($(e.properties.length, r)) === s.True;
}
function ml(e) {
  return J(e, 0);
}
function In(e) {
  const r = Xe();
  return J(e, 0) || J(e, 1) && "length" in e.properties && ce($(e.properties.length, r)) === s.True;
}
function ll(e) {
  const r = Dr([Ar()], Ar());
  return J(e, 0) || J(e, 1) && "then" in e.properties && ce($(e.properties.then, r)) === s.True;
}
function yi(e, r) {
  return $(e, r) === s.False || nn(e) && !nn(r) ? s.False : s.True;
}
function V(e, r) {
  return le(e) ? s.False : de(e) ? s.Union : Te(e) || Yo(e) && fo(r) || Xo(e) && Zn(r) || Qo(e) && lo(r) || Rr(e) && mo(r) || hn(e) && al(r) || me(e) && fo(r) || Rr(e) && mo(r) || G(e) && Zn(r) || we(e) && Zn(r) || Je(e) && lo(r) || Ur(e) && fl(r) || bn(e) && sl(r) || gn(e) && ml(r) || Fn(e) && dl(r) ? s.True : M(e) && me(Bn(e)) ? r[an] === "Record" ? s.True : s.False : M(e) && G(Bn(e)) ? J(r, 0) ? s.True : s.False : s.False;
}
function pl(e, r) {
  return C(r) ? N(e, r) : M(r) ? fe(e, r) : A(r) ? (() => {
    for (const n of Object.getOwnPropertyNames(r.properties)) {
      if (!(n in e.properties) && !nn(r.properties[n]))
        return s.False;
      if (nn(r.properties[n]))
        return s.True;
      if (yi(e.properties[n], r.properties[n]) === s.False)
        return s.False;
    }
    return s.True;
  })() : s.False;
}
function hl(e, r) {
  return C(r) ? N(e, r) : A(r) && ll(r) ? s.True : dt(r) ? ce($(e.item, r.item)) : s.False;
}
function Bn(e) {
  return ir in e.patternProperties ? Xe() : ur in e.patternProperties ? Le() : gr("Unknown record key pattern");
}
function Kn(e) {
  return ir in e.patternProperties ? e.patternProperties[ir] : ur in e.patternProperties ? e.patternProperties[ur] : gr("Unable to get record value schema");
}
function fe(e, r) {
  const [n, t] = [Bn(r), Kn(r)];
  return Yo(e) && G(n) && ce($(e, t)) === s.True ? s.True : Ur(e) && G(n) || me(e) && G(n) || qe(e) && G(n) ? $(e, t) : A(e) ? (() => {
    for (const o of Object.getOwnPropertyNames(e.properties))
      if (yi(t, e.properties[o]) === s.False)
        return s.False;
    return s.True;
  })() : s.False;
}
function gl(e, r) {
  return C(r) ? N(e, r) : A(r) ? V(e, r) : M(r) ? $(Kn(e), Kn(r)) : s.False;
}
function bl(e, r) {
  const n = Ir(e) ? Le() : e, t = Ir(r) ? Le() : r;
  return $(n, t);
}
function ki(e, r) {
  return ze(e) && T(e.const) || me(e) ? s.True : s.False;
}
function Fl(e, r) {
  return C(r) ? N(e, r) : A(r) ? V(e, r) : M(r) ? fe(e, r) : me(r) ? s.True : s.False;
}
function vl(e, r) {
  return C(r) ? N(e, r) : A(r) ? V(e, r) : M(r) ? fe(e, r) : Rr(r) ? s.True : s.False;
}
function Ol(e, r) {
  return xr(e) ? $(tn(e), r) : xr(r) ? $(e, tn(r)) : gr("Invalid fallthrough for TemplateLiteral");
}
function $l(e, r) {
  return qe(r) && e.items !== void 0 && e.items.every((n) => $(n, r.items) === s.True);
}
function _l(e, r) {
  return Te(e) ? s.True : le(e) ? s.False : de(e) ? s.Union : s.False;
}
function wl(e, r) {
  return C(r) ? N(e, r) : A(r) && In(r) || qe(r) && $l(e, r) ? s.True : vn(r) ? S(e.items) && !S(r.items) || !S(e.items) && S(r.items) ? s.False : S(e.items) && !S(r.items) || e.items.every((n, t) => $(n, r.items[t]) === s.True) ? s.True : s.False : s.False;
}
function yl(e, r) {
  return C(r) ? N(e, r) : A(r) ? V(e, r) : M(r) ? fe(e, r) : Ur(r) ? s.True : s.False;
}
function kl(e, r) {
  return C(r) ? N(e, r) : A(r) ? V(e, r) : M(r) ? fe(e, r) : On(r) ? xl(e) : De(r) ? s.True : s.False;
}
function wt(e, r) {
  return r.anyOf.some((n) => $(e, n) === s.True) ? s.True : s.False;
}
function Il(e, r) {
  return e.anyOf.every((n) => $(n, r) === s.True) ? s.True : s.False;
}
function Ii(e, r) {
  return s.True;
}
function Rl(e, r) {
  return Te(r) ? _i() : Ye(r) ? kn(e, r) : $e(r) ? wt(e, r) : de(r) ? _t() : me(r) ? ki(e) : G(r) ? wi(e) : we(r) ? $i(e) : Je(r) ? Oi(e) : qe(r) ? Vm(e) : vn(r) ? _l(e) : A(r) ? V(e, r) : le(r) ? s.True : s.False;
}
function xl(e, r) {
  return De(e) || De(e) ? s.True : s.False;
}
function Al(e, r) {
  return Ye(r) ? kn(e, r) : $e(r) ? wt(e, r) : le(r) ? Ii() : de(r) ? _t() : A(r) ? V(e, r) : On(r) ? s.True : s.False;
}
function $(e, r) {
  return (
    // resolvable
    xr(e) || xr(r) ? Ol(e, r) : Ir(e) || Ir(r) ? bl(e, r) : or(e) || or(r) ? il(e, r) : (
      // standard
      de(e) ? Km(e, r) : qe(e) ? Hm(e, r) : hn(e) ? qm(e, r) : Je(e) ? Jm(e, r) : at(e) ? Gm(e, r) : gn(e) ? Ym(e, r) : bn(e) ? Xm(e, r) : Fn(e) ? Qm(e, r) : we(e) ? el(e, r) : Ye(e) ? rl(e, r) : st(e) ? nl(e, r) : ze(e) ? tl(e, r) : Te(e) ? ol() : ft(e) ? ul(e, r) : G(e) ? cl(e, r) : A(e) ? pl(e, r) : M(e) ? gl(e, r) : me(e) ? Fl(e, r) : Rr(e) ? vl(e, r) : vn(e) ? wl(e, r) : dt(e) ? hl(e, r) : Ur(e) ? yl(e, r) : De(e) ? kl(e, r) : $e(e) ? Il(e, r) : le(e) ? Rl(e, r) : On(e) ? Al(e, r) : gr(`Unknown left type operand '${e[g]}'`)
    )
  );
}
function Mr(e, r) {
  return $(e, r);
}
function zl(e, r, n, t, o) {
  const u = {};
  for (const c of globalThis.Object.getOwnPropertyNames(e))
    u[c] = yt(e[c], r, n, t, re(o));
  return u;
}
function Tl(e, r, n, t, o) {
  return zl(e.properties, r, n, t, o);
}
function Pl(e, r, n, t, o) {
  const u = Tl(e, r, n, t, o);
  return B(u);
}
function Sl(e, r, n, t) {
  const o = Mr(e, r);
  return o === s.Union ? K([n, t]) : o === s.True ? n : t;
}
function yt(e, r, n, t, o) {
  return oe(e) ? Pl(e, r, n, t, o) : Ve(e) ? m(Cl(e, r, n, t, o)) : m(Sl(e, r, n, t), o);
}
function Zl(e, r, n, t, o) {
  return {
    [e]: yt(Z(e), r, n, t, re(o))
  };
}
function jl(e, r, n, t, o) {
  return e.reduce((u, c) => ({ ...u, ...Zl(c, r, n, t, o) }), {});
}
function El(e, r, n, t, o) {
  return jl(e.keys, r, n, t, o);
}
function Cl(e, r, n, t, o) {
  const u = El(e, r, n, t, o);
  return B(u);
}
function Nl(e, r) {
  return kt(tn(e), r);
}
function Ul(e, r) {
  const n = e.filter((t) => Mr(t, r) === s.False);
  return n.length === 1 ? n[0] : K(n);
}
function kt(e, r, n = {}) {
  return He(e) ? m(Nl(e, r), n) : oe(e) ? m(Ml(e, r), n) : m(L(e) ? Ul(e.anyOf, r) : Mr(e, r) !== s.False ? E() : e, n);
}
function Dl(e, r) {
  const n = {};
  for (const t of globalThis.Object.getOwnPropertyNames(e))
    n[t] = kt(e[t], r);
  return n;
}
function Ll(e, r) {
  return Dl(e.properties, r);
}
function Ml(e, r) {
  const n = Ll(e, r);
  return B(n);
}
function Wl(e, r) {
  return It(tn(e), r);
}
function Bl(e, r) {
  const n = e.filter((t) => Mr(t, r) !== s.False);
  return n.length === 1 ? n[0] : K(n);
}
function It(e, r, n) {
  return He(e) ? m(Wl(e, r), n) : oe(e) ? m(Hl(e, r), n) : m(L(e) ? Bl(e.anyOf, r) : Mr(e, r) !== s.False ? e : E(), n);
}
function Kl(e, r) {
  const n = {};
  for (const t of globalThis.Object.getOwnPropertyNames(e))
    n[t] = It(e[t], r);
  return n;
}
function Vl(e, r) {
  return Kl(e.properties, r);
}
function Hl(e, r) {
  const n = Vl(e, r);
  return B(n);
}
function Gl(e, r) {
  return fr(e) ? m(e.returns, r) : E(r);
}
function Ri(e) {
  return Se(Ze(e));
}
function Qe(e, r, n) {
  return m({ [g]: "Record", type: "object", patternProperties: { [e]: r } }, n);
}
function Rt(e, r, n) {
  const t = {};
  for (const o of e)
    t[o] = r;
  return D(t, { ...n, [an]: "Record" });
}
function ql(e, r, n) {
  return vd(e) ? Rt(Pe(e), r, n) : Qe(e.pattern, r, n);
}
function Jl(e, r, n) {
  return Rt(Pe(K(e)), r, n);
}
function Yl(e, r, n) {
  return Rt([e.toString()], r, n);
}
function Xl(e, r, n) {
  return Qe(e.source, r, n);
}
function Ql(e, r, n) {
  const t = S(e.pattern) ? ur : e.pattern;
  return Qe(t, r, n);
}
function ep(e, r, n) {
  return Qe(ur, r, n);
}
function rp(e, r, n) {
  return Qe(Jf, r, n);
}
function np(e, r, n) {
  return D({ true: r, false: r }, n);
}
function tp(e, r, n) {
  return Qe(ir, r, n);
}
function op(e, r, n) {
  return Qe(ir, r, n);
}
function xi(e, r, n = {}) {
  return L(e) ? Jl(e.anyOf, r, n) : He(e) ? ql(e, r, n) : Ke(e) ? Yl(e.const, r, n) : Er(e) ? np(e, r, n) : mr(e) ? tp(e, r, n) : lr(e) ? op(e, r, n) : Vo(e) ? Xl(e, r, n) : Nr(e) ? Ql(e, r, n) : Wo(e) ? ep(e, r, n) : Cr(e) ? rp(e, r, n) : E(n);
}
function xt(e) {
  return globalThis.Object.getOwnPropertyNames(e.patternProperties)[0];
}
function ip(e) {
  const r = xt(e);
  return r === ur ? Le() : r === ir ? Xe() : Le({ pattern: r });
}
function Ai(e) {
  return e.patternProperties[xt(e)];
}
function up(e, r) {
  return r.parameters = Wr(e, r.parameters), r.returns = pe(e, r.returns), r;
}
function cp(e, r) {
  return r.parameters = Wr(e, r.parameters), r.returns = pe(e, r.returns), r;
}
function ap(e, r) {
  return r.allOf = Wr(e, r.allOf), r;
}
function sp(e, r) {
  return r.anyOf = Wr(e, r.anyOf), r;
}
function fp(e, r) {
  return S(r.items) || (r.items = Wr(e, r.items)), r;
}
function dp(e, r) {
  return r.items = pe(e, r.items), r;
}
function mp(e, r) {
  return r.items = pe(e, r.items), r;
}
function lp(e, r) {
  return r.items = pe(e, r.items), r;
}
function pp(e, r) {
  return r.item = pe(e, r.item), r;
}
function hp(e, r) {
  const n = vp(e, r.properties);
  return { ...r, ...D(n) };
}
function gp(e, r) {
  const n = pe(e, ip(r)), t = pe(e, Ai(r)), o = xi(n, t);
  return { ...r, ...o };
}
function bp(e, r) {
  return r.index in e ? e[r.index] : yn();
}
function Fp(e, r) {
  const n = nt(r), t = Ae(r), o = pe(e, r);
  return n && t ? Ri(o) : n && !t ? Se(o) : !n && t ? Ze(o) : o;
}
function vp(e, r) {
  return globalThis.Object.getOwnPropertyNames(r).reduce((n, t) => ({ ...n, [t]: Fp(e, r[t]) }), {});
}
function Wr(e, r) {
  return r.map((n) => pe(e, n));
}
function pe(e, r) {
  return fr(r) ? up(e, r) : dr(r) ? cp(e, r) : se(r) ? ap(e, r) : L(r) ? sp(e, r) : Ge(r) ? fp(e, r) : ar(r) ? dp(e, r) : sn(r) ? mp(e, r) : dn(r) ? lp(e, r) : mn(r) ? pp(e, r) : he(r) ? hp(e, r) : ln(r) ? gp(e, r) : Bo(r) ? bp(e, r) : r;
}
function Op(e, r) {
  return pe(r, rt(e));
}
function $p(e) {
  return m({ [g]: "Integer", type: "integer" }, e);
}
function _p(e, r, n) {
  return {
    [e]: br(Z(e), r, re(n))
  };
}
function wp(e, r, n) {
  return e.reduce((o, u) => ({ ...o, ..._p(u, r, n) }), {});
}
function yp(e, r, n) {
  return wp(e.keys, r, n);
}
function kp(e, r, n) {
  const t = yp(e, r, n);
  return B(t);
}
function Ip(e) {
  const [r, n] = [e.slice(0, 1), e.slice(1)];
  return [r.toLowerCase(), n].join("");
}
function Rp(e) {
  const [r, n] = [e.slice(0, 1), e.slice(1)];
  return [r.toUpperCase(), n].join("");
}
function xp(e) {
  return e.toUpperCase();
}
function Ap(e) {
  return e.toLowerCase();
}
function zp(e, r, n) {
  const t = gt(e.pattern);
  if (!Tr(t))
    return { ...e, pattern: zi(e.pattern, r) };
  const c = [...$n(t)].map((p) => Z(p)), a = Ti(c, r), l = K(a);
  return ui([l], n);
}
function zi(e, r) {
  return typeof e == "string" ? r === "Uncapitalize" ? Ip(e) : r === "Capitalize" ? Rp(e) : r === "Uppercase" ? xp(e) : r === "Lowercase" ? Ap(e) : e : e.toString();
}
function Ti(e, r) {
  return e.map((n) => br(n, r));
}
function br(e, r, n = {}) {
  return (
    // Intrinsic-Mapped-Inference
    Ve(e) ? kp(e, r, n) : (
      // Standard-Inference
      He(e) ? zp(e, r, n) : L(e) ? K(Ti(e.anyOf, r), n) : Ke(e) ? Z(zi(e.const, r), n) : (
        // Default Type
        m(e, n)
      )
    )
  );
}
function Tp(e, r = {}) {
  return br(e, "Capitalize", r);
}
function Pp(e, r = {}) {
  return br(e, "Lowercase", r);
}
function Sp(e, r = {}) {
  return br(e, "Uncapitalize", r);
}
function Zp(e, r = {}) {
  return br(e, "Uppercase", r);
}
function jp(e, r, n) {
  const t = {};
  for (const o of globalThis.Object.getOwnPropertyNames(e))
    t[o] = Rn(e[o], r, re(n));
  return t;
}
function Ep(e, r, n) {
  return jp(e.properties, r, n);
}
function Cp(e, r, n) {
  const t = Ep(e, r, n);
  return B(t);
}
function Np(e, r) {
  return e.map((n) => At(n, r));
}
function Up(e, r) {
  return e.map((n) => At(n, r));
}
function Dp(e, r) {
  const { [r]: n, ...t } = e;
  return t;
}
function Lp(e, r) {
  return r.reduce((n, t) => Dp(n, t), e);
}
function Mp(e, r) {
  const n = ne(e, [ue, "$id", "required", "properties"]), t = Lp(e.properties, r);
  return D(t, n);
}
function Wp(e) {
  const r = e.reduce((n, t) => Ko(t) ? [...n, Z(t)] : n, []);
  return K(r);
}
function At(e, r) {
  return se(e) ? je(Np(e.allOf, r)) : L(e) ? K(Up(e.anyOf, r)) : he(e) ? Mp(e, r) : D({});
}
function Rn(e, r, n) {
  const t = X(r) ? Wp(r) : r, o = Re(r) ? Pe(r) : r, u = q(e), c = q(r);
  return oe(e) ? Cp(e, o, n) : Ve(r) ? Hp(e, r, n) : u && c ? U("Omit", [e, t], n) : !u && c ? U("Omit", [e, t], n) : u && !c ? U("Omit", [e, t], n) : m({ ...At(e, o), ...n });
}
function Bp(e, r, n) {
  return { [r]: Rn(e, [r], re(n)) };
}
function Kp(e, r, n) {
  return r.reduce((t, o) => ({ ...t, ...Bp(e, o, n) }), {});
}
function Vp(e, r, n) {
  return Kp(e, r.keys, n);
}
function Hp(e, r, n) {
  const t = Vp(e, r, n);
  return B(t);
}
function Gp(e, r, n) {
  const t = {};
  for (const o of globalThis.Object.getOwnPropertyNames(e))
    t[o] = xn(e[o], r, re(n));
  return t;
}
function qp(e, r, n) {
  return Gp(e.properties, r, n);
}
function Jp(e, r, n) {
  const t = qp(e, r, n);
  return B(t);
}
function Yp(e, r) {
  return e.map((n) => zt(n, r));
}
function Xp(e, r) {
  return e.map((n) => zt(n, r));
}
function Qp(e, r) {
  const n = {};
  for (const t of r)
    t in e && (n[t] = e[t]);
  return n;
}
function eh(e, r) {
  const n = ne(e, [ue, "$id", "required", "properties"]), t = Qp(e.properties, r);
  return D(t, n);
}
function rh(e) {
  const r = e.reduce((n, t) => Ko(t) ? [...n, Z(t)] : n, []);
  return K(r);
}
function zt(e, r) {
  return se(e) ? je(Yp(e.allOf, r)) : L(e) ? K(Xp(e.anyOf, r)) : he(e) ? eh(e, r) : D({});
}
function xn(e, r, n) {
  const t = X(r) ? rh(r) : r, o = Re(r) ? Pe(r) : r, u = q(e), c = q(r);
  return oe(e) ? Jp(e, o, n) : Ve(r) ? ih(e, r, n) : u && c ? U("Pick", [e, t], n) : !u && c ? U("Pick", [e, t], n) : u && !c ? U("Pick", [e, t], n) : m({ ...zt(e, o), ...n });
}
function nh(e, r, n) {
  return {
    [r]: xn(e, [r], re(n))
  };
}
function th(e, r, n) {
  return r.reduce((t, o) => ({ ...t, ...nh(e, o, n) }), {});
}
function oh(e, r, n) {
  return th(e, r.keys, n);
}
function ih(e, r, n) {
  const t = oh(e, r, n);
  return B(t);
}
function uh(e, r) {
  return U("Partial", [U(e, r)]);
}
function ch(e) {
  return U("Partial", [Lr(e)]);
}
function ah(e) {
  const r = {};
  for (const n of globalThis.Object.getOwnPropertyNames(e))
    r[n] = Ze(e[n]);
  return r;
}
function sh(e) {
  const r = ne(e, [ue, "$id", "required", "properties"]), n = ah(e.properties);
  return D(n, r);
}
function po(e) {
  return e.map((r) => Pi(r));
}
function Pi(e) {
  return (
    // Mappable
    sr(e) ? uh(e.target, e.parameters) : q(e) ? ch(e.$ref) : se(e) ? je(po(e.allOf)) : L(e) ? K(po(e.anyOf)) : he(e) ? sh(e) : (
      // Intrinsic
      fn(e) || Er(e) || mr(e) || Ke(e) || tt(e) || lr(e) || Nr(e) || ot(e) || it(e) ? e : (
        // Passthrough
        D({})
      )
    )
  );
}
function Tt(e, r) {
  return oe(e) ? mh(e, r) : m({ ...Pi(e), ...r });
}
function fh(e, r) {
  const n = {};
  for (const t of globalThis.Object.getOwnPropertyNames(e))
    n[t] = Tt(e[t], re(r));
  return n;
}
function dh(e, r) {
  return fh(e.properties, r);
}
function mh(e, r) {
  const n = dh(e, r);
  return B(n);
}
function lh(e, r) {
  return U("Required", [U(e, r)]);
}
function ph(e) {
  return U("Required", [Lr(e)]);
}
function hh(e) {
  const r = {};
  for (const n of globalThis.Object.getOwnPropertyNames(e))
    r[n] = ne(e[n], [_e]);
  return r;
}
function gh(e) {
  const r = ne(e, [ue, "$id", "required", "properties"]), n = hh(e.properties);
  return D(n, r);
}
function ho(e) {
  return e.map((r) => Si(r));
}
function Si(e) {
  return (
    // Mappable
    sr(e) ? lh(e.target, e.parameters) : q(e) ? ph(e.$ref) : se(e) ? je(ho(e.allOf)) : L(e) ? K(ho(e.anyOf)) : he(e) ? gh(e) : (
      // Intrinsic
      fn(e) || Er(e) || mr(e) || Ke(e) || tt(e) || lr(e) || Nr(e) || ot(e) || it(e) ? e : (
        // Passthrough
        D({})
      )
    )
  );
}
function Pt(e, r) {
  return oe(e) ? vh(e, r) : m({ ...Si(e), ...r });
}
function bh(e, r) {
  const n = {};
  for (const t of globalThis.Object.getOwnPropertyNames(e))
    n[t] = Pt(e[t], r);
  return n;
}
function Fh(e, r) {
  return bh(e.properties, r);
}
function vh(e, r) {
  const n = Fh(e, r);
  return B(n);
}
function Oh(e, r) {
  return r.map((n) => q(n) ? St(e, n.$ref) : te(e, n));
}
function St(e, r) {
  return r in e ? q(e[r]) ? St(e, e[r].$ref) : te(e, e[r]) : E();
}
function $h(e) {
  return wn(e[0]);
}
function _h(e) {
  return _n(e[0], e[1]);
}
function wh(e) {
  return Ot(e[0]);
}
function yh(e) {
  return Tt(e[0]);
}
function kh(e) {
  return Rn(e[0], e[1]);
}
function Ih(e) {
  return xn(e[0], e[1]);
}
function Rh(e) {
  return Pt(e[0]);
}
function xh(e, r, n) {
  const t = Oh(e, n);
  return r === "Awaited" ? $h(t) : r === "Index" ? _h(t) : r === "KeyOf" ? wh(t) : r === "Partial" ? yh(t) : r === "Omit" ? kh(t) : r === "Pick" ? Ih(t) : r === "Required" ? Rh(t) : E();
}
function Ah(e, r) {
  return mt(te(e, r));
}
function zh(e, r) {
  return lt(te(e, r));
}
function Th(e, r, n) {
  return pt(Br(e, r), te(e, n));
}
function Ph(e, r, n) {
  return Dr(Br(e, r), te(e, n));
}
function Sh(e, r) {
  return je(Br(e, r));
}
function Zh(e, r) {
  return Ft(te(e, r));
}
function jh(e, r) {
  return D(globalThis.Object.keys(r).reduce((n, t) => ({ ...n, [t]: te(e, r[t]) }), {}));
}
function Eh(e, r) {
  const [n, t] = [te(e, Ai(r)), xt(r)], o = rt(r);
  return o.patternProperties[t] = n, o;
}
function Ch(e, r) {
  return q(r) ? { ...St(e, r.$ref), [ue]: r[ue] } : r;
}
function Nh(e, r) {
  return hr(Br(e, r));
}
function Uh(e, r) {
  return K(Br(e, r));
}
function Br(e, r) {
  return r.map((n) => te(e, n));
}
function te(e, r) {
  return (
    // Modifiers
    Ae(r) ? m(te(e, ne(r, [_e])), r) : nt(r) ? m(te(e, ne(r, [jr])), r) : (
      // Transform
      pn(r) ? m(Ch(e, r), r) : (
        // Types
        ar(r) ? m(Ah(e, r.items), r) : sn(r) ? m(zh(e, r.items), r) : sr(r) ? m(xh(e, r.target, r.parameters)) : fr(r) ? m(Th(e, r.parameters, r.returns), r) : dr(r) ? m(Ph(e, r.parameters, r.returns), r) : se(r) ? m(Sh(e, r.allOf), r) : dn(r) ? m(Zh(e, r.items), r) : he(r) ? m(jh(e, r.properties), r) : ln(r) ? m(Eh(e, r)) : Ge(r) ? m(Nh(e, r.items || []), r) : L(r) ? m(Uh(e, r.anyOf), r) : r
      )
    )
  );
}
function Dh(e, r) {
  return r in e ? te(e, e[r]) : E();
}
function Lh(e) {
  return globalThis.Object.getOwnPropertyNames(e).reduce((r, n) => ({ ...r, [n]: Dh(e, n) }), {});
}
class Mh {
  constructor(r) {
    const n = Lh(r), t = this.WithIdentifiers(n);
    this.$defs = t;
  }
  /** `[Json]` Imports a Type by Key. */
  Import(r, n) {
    const t = { ...this.$defs, [r]: m(this.$defs[r], n) };
    return m({ [g]: "Import", $defs: t, $ref: r });
  }
  // prettier-ignore
  WithIdentifiers(r) {
    return globalThis.Object.getOwnPropertyNames(r).reduce((n, t) => ({ ...n, [t]: { ...r[t], $id: t } }), {});
  }
}
function Wh(e) {
  return new Mh(e);
}
function Bh(e, r) {
  return m({ [g]: "Not", not: e }, r);
}
function Kh(e, r) {
  return dr(e) ? hr(e.parameters, r) : E();
}
let Vh = 0;
function Hh(e, r = {}) {
  S(r.$id) && (r.$id = `T${Vh++}`);
  const n = rt(e({ [g]: "This", $ref: `${r.$id}` }));
  return n.$id = r.$id, m({ [an]: "Recursive", ...n }, r);
}
function Gh(e, r) {
  const n = T(e) ? new globalThis.RegExp(e) : e;
  return m({ [g]: "RegExp", type: "RegExp", source: n.source, flags: n.flags }, r);
}
function qh(e) {
  return se(e) ? e.allOf : L(e) ? e.anyOf : Ge(e) ? e.items ?? [] : [];
}
function Jh(e) {
  return qh(e);
}
function Yh(e, r) {
  return dr(e) ? m(e.returns, r) : E(r);
}
class Xh {
  constructor(r) {
    this.schema = r;
  }
  Decode(r) {
    return new Qh(this.schema, r);
  }
}
class Qh {
  constructor(r, n) {
    this.schema = r, this.decode = n;
  }
  EncodeTransform(r, n) {
    const u = { Encode: (c) => n[ue].Encode(r(c)), Decode: (c) => this.decode(n[ue].Decode(c)) };
    return { ...n, [ue]: u };
  }
  EncodeSchema(r, n) {
    const t = { Decode: this.decode, Encode: r };
    return { ...n, [ue]: t };
  }
  Encode(r) {
    return pn(this.schema) ? this.EncodeTransform(r, this.schema) : this.EncodeSchema(r, this.schema);
  }
}
function eg(e) {
  return new Xh(e);
}
function rg(e = {}) {
  return m({ [g]: e[g] ?? "Unsafe" }, e);
}
function ng(e) {
  return m({ [g]: "Void", type: "void" }, e);
}
const tg = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  Any: Ar,
  Argument: td,
  Array: mt,
  AsyncIterator: lt,
  Awaited: wn,
  BigInt: bt,
  Boolean: oi,
  Capitalize: Tp,
  Composite: Nm,
  Const: Lm,
  Constructor: pt,
  ConstructorParameters: Mm,
  Date: hi,
  Enum: Wm,
  Exclude: kt,
  Extends: yt,
  Extract: It,
  Function: Dr,
  Index: _n,
  InstanceType: Gl,
  Instantiate: Op,
  Integer: $p,
  Intersect: je,
  Iterator: Ft,
  KeyOf: Ot,
  Literal: Z,
  Lowercase: Pp,
  Mapped: um,
  Module: Wh,
  Never: E,
  Not: Bh,
  Null: gi,
  Number: Xe,
  Object: D,
  Omit: Rn,
  Optional: Ze,
  Parameters: Kh,
  Partial: Tt,
  Pick: xn,
  Promise: fi,
  Readonly: Se,
  ReadonlyOptional: Ri,
  Record: xi,
  Recursive: Hh,
  Ref: Lr,
  RegExp: Gh,
  Required: Pt,
  Rest: Jh,
  ReturnType: Yh,
  String: Le,
  Symbol: bi,
  TemplateLiteral: ui,
  Transform: eg,
  Tuple: hr,
  Uint8Array: vi,
  Uncapitalize: Sp,
  Undefined: Fi,
  Union: K,
  Unknown: yn,
  Unsafe: rg,
  Uppercase: Zp,
  Void: ng
}, Symbol.toStringTag, { value: "Module" })), i = tg, d = i.String({ $id: "Color" }), H = i.String({ $id: "FamilyName" }), ve = i.String({ $id: "FontWeight" }), ke = i.String({ $id: "Length" }), Ie = i.String({ $id: "Percentage" }), _r = i.String({ $id: "BoxShadow" }), og = i.String({ $id: "Number" }), ig = i.String({ $id: "Size" }), ye = i.Union([i.String(), i.Literal("thin"), i.Literal("medium"), i.Literal("thick")], {
  $id: "LineWidth"
}), jn = i.Optional(i.Object({
  columnGap: i.Optional(i.Union([i.Ref(ke), i.Ref(Ie)])),
  rowGap: i.Optional(i.Union([i.Ref(ke), i.Ref(Ie)])),
  field: i.Optional(i.Object({
    label: i.Optional(i.Object({
      foreground: i.Optional(i.Ref(d)),
      fontFamily: i.Optional(i.Ref(H)),
      fontWeight: i.Optional(i.Ref(ve))
    })),
    input: i.Optional(i.Object({
      background: i.Optional(i.Ref(d)),
      backgroundSubdued: i.Optional(i.Ref(d)),
      foreground: i.Optional(i.Ref(d)),
      foregroundSubdued: i.Optional(i.Ref(d)),
      borderColor: i.Optional(i.Ref(d)),
      borderColorHover: i.Optional(i.Ref(d)),
      borderColorFocus: i.Optional(i.Ref(d)),
      boxShadow: i.Optional(i.Ref(_r)),
      boxShadowHover: i.Optional(i.Ref(_r)),
      boxShadowFocus: i.Optional(i.Ref(_r)),
      height: i.Optional(i.Ref(ig)),
      padding: i.Optional(i.Union([i.Ref(ke), i.Ref(Ie)]))
    }))
  }))
})), ug = i.Object({
  //////////////////////////////////////////////////////////////////////////////////////////////////////////
  // Base border styles
  borderRadius: i.Optional(i.Union([i.Ref(ke), i.Ref(Ie)])),
  borderWidth: i.Optional(i.Ref(ye)),
  //////////////////////////////////////////////////////////////////////////////////////////////////////////
  // Base color palette
  foreground: i.Optional(i.Ref(d)),
  foregroundSubdued: i.Optional(i.Ref(d)),
  foregroundAccent: i.Optional(i.Ref(d)),
  background: i.Optional(i.Ref(d)),
  backgroundNormal: i.Optional(i.Ref(d)),
  backgroundAccent: i.Optional(i.Ref(d)),
  backgroundSubdued: i.Optional(i.Ref(d)),
  borderColor: i.Optional(i.Ref(d)),
  borderColorAccent: i.Optional(i.Ref(d)),
  borderColorSubdued: i.Optional(i.Ref(d)),
  primary: i.Optional(i.Ref(d)),
  primaryBackground: i.Optional(i.Ref(d)),
  primarySubdued: i.Optional(i.Ref(d)),
  primaryAccent: i.Optional(i.Ref(d)),
  secondary: i.Optional(i.Ref(d)),
  secondaryBackground: i.Optional(i.Ref(d)),
  secondarySubdued: i.Optional(i.Ref(d)),
  secondaryAccent: i.Optional(i.Ref(d)),
  success: i.Optional(i.Ref(d)),
  successBackground: i.Optional(i.Ref(d)),
  successSubdued: i.Optional(i.Ref(d)),
  successAccent: i.Optional(i.Ref(d)),
  warning: i.Optional(i.Ref(d)),
  warningBackground: i.Optional(i.Ref(d)),
  warningSubdued: i.Optional(i.Ref(d)),
  warningAccent: i.Optional(i.Ref(d)),
  danger: i.Optional(i.Ref(d)),
  dangerBackground: i.Optional(i.Ref(d)),
  dangerSubdued: i.Optional(i.Ref(d)),
  dangerAccent: i.Optional(i.Ref(d)),
  //////////////////////////////////////////////////////////////////////////////////////////////////////////
  // Base fonts
  fonts: i.Optional(i.Object({
    display: i.Optional(i.Object({
      fontFamily: i.Optional(i.Ref(H)),
      fontWeight: i.Optional(i.Ref(ve))
    })),
    sans: i.Optional(i.Object({
      fontFamily: i.Optional(i.Ref(H)),
      fontWeight: i.Optional(i.Ref(ve))
    })),
    serif: i.Optional(i.Object({
      fontFamily: i.Optional(i.Ref(H)),
      fontWeight: i.Optional(i.Ref(ve))
    })),
    monospace: i.Optional(i.Object({
      fontFamily: i.Optional(i.Ref(H)),
      fontWeight: i.Optional(i.Ref(ve))
    }))
  })),
  //////////////////////////////////////////////////////////////////////////////////////////////////////////
  // Scopes
  navigation: i.Optional(i.Object({
    background: i.Optional(i.Ref(d)),
    backgroundAccent: i.Optional(i.Ref(d)),
    borderWidth: i.Optional(i.Ref(ye)),
    borderColor: i.Optional(i.Ref(d)),
    project: i.Optional(i.Object({
      background: i.Optional(i.Ref(d)),
      foreground: i.Optional(i.Ref(d)),
      fontFamily: i.Optional(i.Ref(H)),
      borderWidth: i.Optional(i.Ref(ye)),
      borderColor: i.Optional(i.Ref(d))
    })),
    modules: i.Optional(i.Object({
      background: i.Optional(i.Ref(d)),
      borderWidth: i.Optional(i.Ref(ye)),
      borderColor: i.Optional(i.Ref(d)),
      button: i.Optional(i.Object({
        foreground: i.Optional(i.Ref(d)),
        foregroundHover: i.Optional(i.Ref(d)),
        foregroundActive: i.Optional(i.Ref(d)),
        background: i.Optional(i.Ref(d)),
        backgroundHover: i.Optional(i.Ref(d)),
        backgroundActive: i.Optional(i.Ref(d))
      }))
    })),
    list: i.Optional(i.Object({
      icon: i.Optional(i.Object({
        foreground: i.Optional(i.Ref(d)),
        foregroundHover: i.Optional(i.Ref(d)),
        foregroundActive: i.Optional(i.Ref(d))
      })),
      foreground: i.Optional(i.Ref(d)),
      foregroundHover: i.Optional(i.Ref(d)),
      foregroundActive: i.Optional(i.Ref(d)),
      background: i.Optional(i.Ref(d)),
      backgroundHover: i.Optional(i.Ref(d)),
      backgroundActive: i.Optional(i.Ref(d)),
      fontFamily: i.Optional(i.Ref(H)),
      divider: i.Object({
        borderColor: i.Optional(i.Ref(d)),
        borderWidth: i.Optional(i.Ref(ye))
      })
    }))
  })),
  header: i.Optional(i.Object({
    background: i.Optional(i.Ref(d)),
    borderWidth: i.Optional(i.Ref(ye)),
    borderColor: i.Optional(i.Ref(d)),
    boxShadow: i.Optional(i.Ref(_r)),
    headline: i.Optional(i.Object({
      foreground: i.Optional(i.Ref(d)),
      fontFamily: i.Optional(i.Ref(H))
    })),
    title: i.Optional(i.Object({
      foreground: i.Optional(i.Ref(d)),
      fontFamily: i.Optional(i.Ref(H)),
      fontWeight: i.Optional(i.Ref(ve))
    }))
  })),
  form: jn,
  sidebar: i.Optional(i.Object({
    background: i.Optional(i.Ref(d)),
    foreground: i.Optional(i.Ref(d)),
    fontFamily: i.Optional(i.Ref(H)),
    borderWidth: i.Optional(i.Ref(ye)),
    borderColor: i.Optional(i.Ref(d)),
    section: i.Optional(i.Object({
      toggle: i.Optional(i.Object({
        icon: i.Optional(i.Object({
          foreground: i.Optional(i.Ref(d)),
          foregroundHover: i.Optional(i.Ref(d)),
          foregroundActive: i.Optional(i.Ref(d))
        })),
        foreground: i.Optional(i.Ref(d)),
        foregroundHover: i.Optional(i.Ref(d)),
        foregroundActive: i.Optional(i.Ref(d)),
        background: i.Optional(i.Ref(d)),
        backgroundHover: i.Optional(i.Ref(d)),
        backgroundActive: i.Optional(i.Ref(d)),
        fontFamily: i.Optional(i.Ref(H)),
        borderWidth: i.Optional(i.Ref(ye)),
        borderColor: i.Optional(i.Ref(d))
      })),
      form: jn
    }))
  })),
  public: i.Optional(i.Object({
    background: i.Optional(i.Ref(d)),
    foreground: i.Optional(i.Ref(d)),
    foregroundAccent: i.Optional(i.Ref(d)),
    art: i.Optional(i.Object({
      background: i.Optional(i.Ref(d)),
      primary: i.Optional(i.Ref(d)),
      secondary: i.Optional(i.Ref(d)),
      speed: i.Optional(i.Ref(og))
    })),
    form: jn
  })),
  popover: i.Optional(i.Object({
    menu: i.Optional(i.Object({
      background: i.Optional(i.Ref(d)),
      borderRadius: i.Optional(i.Optional(i.Union([i.Ref(ke), i.Ref(Ie)]))),
      boxShadow: i.Optional(i.Ref(_r))
    }))
  })),
  banner: i.Optional(i.Object({
    background: i.Optional(i.Ref(d)),
    padding: i.Optional(i.Union([i.Ref(ke), i.Ref(Ie)])),
    borderRadius: i.Optional(i.Optional(i.Union([i.Ref(ke), i.Ref(Ie)]))),
    avatar: i.Optional(i.Object({
      background: i.Optional(i.Ref(d)),
      foreground: i.Optional(i.Ref(d)),
      borderRadius: i.Optional(i.Optional(i.Union([i.Ref(ke), i.Ref(Ie)])))
    })),
    headline: i.Optional(i.Object({
      foreground: i.Optional(i.Ref(d)),
      fontFamily: i.Optional(i.Ref(H)),
      fontWeight: i.Optional(i.Ref(ve))
    })),
    title: i.Optional(i.Object({
      foreground: i.Optional(i.Ref(d)),
      fontFamily: i.Optional(i.Ref(H)),
      fontWeight: i.Optional(i.Ref(ve))
    })),
    subtitle: i.Optional(i.Object({
      foreground: i.Optional(i.Ref(d)),
      fontFamily: i.Optional(i.Ref(H)),
      fontWeight: i.Optional(i.Ref(ve))
    })),
    art: i.Optional(i.Object({
      foreground: i.Optional(i.Ref(d))
    }))
  }))
}), cg = i.Object({
  id: i.String(),
  name: i.String(),
  appearance: i.Union([i.Literal("light"), i.Literal("dark")]),
  rules: ug
});
var go;
(function(e) {
  e[e.None = 0] = "None", e[e.RemainingAdmins = 1] = "RemainingAdmins", e[e.UserLimits = 2] = "UserLimits", e[e.All = 3] = "All";
})(go || (go = {}));
const ag = Qn([kr(), Ss()]);
Ee({
  type: kr(),
  uid: ag.optional()
}).passthrough();
const sg = (e) => {
  const r = Ce(() => {
    const o = /* @__PURE__ */ new Map(), u = (c, a = []) => {
      for (const [l, p] of Object.entries(c))
        typeof p == "object" && p !== null && ("type" in p && p.type === "object" && "properties" in p && u(p.properties, [...a, l]), "$ref" in p && p.$ref === "FamilyName" && (o.has(a) ? o.set(a, { family: l, weight: o.get(a).weight }) : o.set(a, { family: l, weight: null })), "$ref" in p && p.$ref === "FontWeight" && (o.has(a) ? o.set(a, { family: o.get(a).family, weight: l }) : o.set(a, { family: null, weight: l })));
    };
    return u(cg.properties.rules.properties), o;
  }), n = Ce(() => {
    const o = /* @__PURE__ */ new Map();
    for (const [u, { family: c, weight: a }] of r.value.entries()) {
      let l = null, p = null;
      if (c && (l = Ct(ee(e).rules, [...u, c])), a && (p = Ct(ee(e).rules, [...u, a])), l) {
        const b = l.split(",");
        for (const y of b) {
          const O = y.trim();
          if (O.startsWith("var(--")) {
            b.push(Ji(O.slice(6, -1)));
            continue;
          }
          if ((O.startsWith('"') && O.endsWith('"')) === !1)
            continue;
          const I = O.slice(1, -1);
          o.has(I) ? o.get(I).add(p ?? "400") : o.set(I, /* @__PURE__ */ new Set([p ?? "400"]));
        }
      }
    }
    return o;
  });
  return { googleFonts: Ce(() => {
    const o = [];
    for (const [u, c] of n.value.entries())
      if (["Inter", "Merriweather", "Fira Mono"].includes(u) === !1) {
        const l = Array.from(c).sort((p, b) => Number(p) - Number(b)).join(";");
        o.push(`${u.replaceAll(" ", "+")}:wght@${l}`);
      }
    return o;
  }) };
}, An = (e) => e, Zt = An({
  id: "Directus Default",
  name: "$t:theme_directus_default",
  appearance: "dark",
  rules: {
    borderRadius: "6px",
    borderWidth: "2px",
    foreground: "#c9d1d9",
    foregroundAccent: "#f0f6fc",
    foregroundSubdued: "#666672",
    background: "#0d1117",
    backgroundNormal: "#21262e",
    backgroundAccent: "#30363d",
    backgroundSubdued: "#161b22",
    borderColor: "#21262e",
    borderColorAccent: "#30363d",
    borderColorSubdued: "#21262d",
    primary: "var(--project-color)",
    primaryBackground: "color-mix(in srgb, var(--theme--background), var(--theme--primary) 10%)",
    primarySubdued: "color-mix(in srgb, var(--theme--background), var(--theme--primary) 50%)",
    primaryAccent: "color-mix(in srgb, var(--theme--primary), #16151a 25%)",
    secondary: "#ff99dd",
    secondaryBackground: "color-mix(in srgb, var(--theme--background), var(--theme--secondary) 10%)",
    secondarySubdued: "color-mix(in srgb, var(--theme--background), var(--theme--secondary) 50%)",
    secondaryAccent: "color-mix(in srgb, var(--theme--secondary), #16151a 25%)",
    success: "#2ecda7",
    successBackground: "color-mix(in srgb, var(--theme--background), var(--theme--success) 10%)",
    successSubdued: "color-mix(in srgb, var(--theme--background), var(--theme--success) 50%)",
    successAccent: "color-mix(in srgb, var(--theme--success), #16151a 25%)",
    warning: "#ffa439",
    warningBackground: "color-mix(in srgb, var(--theme--background), var(--theme--warning) 10%)",
    warningSubdued: "color-mix(in srgb, var(--theme--background), var(--theme--warning) 50%)",
    warningAccent: "color-mix(in srgb, var(--theme--warning), #16151a 25%)",
    danger: "#e35169",
    dangerBackground: "color-mix(in srgb, var(--theme--background), var(--theme--danger) 10%)",
    dangerSubdued: "color-mix(in srgb, var(--theme--background), var(--theme--danger) 50%)",
    dangerAccent: "color-mix(in srgb, var(--theme--danger), #16151a 25%)",
    fonts: {
      display: {
        fontFamily: '"Inter", system-ui',
        fontWeight: "700"
      },
      sans: {
        fontFamily: '"Inter", system-ui',
        fontWeight: "500"
      },
      serif: {
        fontFamily: '"Merriweather", serif',
        fontWeight: "500"
      },
      monospace: {
        fontFamily: '"Fira Mono", monospace',
        fontWeight: "500"
      }
    },
    navigation: {
      background: "#21262e",
      backgroundAccent: "#30363d",
      borderColor: "transparent",
      borderWidth: "0px",
      project: {
        borderColor: "transparent",
        borderWidth: "0px",
        background: "#30363d",
        foreground: "var(--theme--foreground-accent)",
        fontFamily: "var(--theme--fonts--sans--font-family)"
      },
      modules: {
        background: "var(--theme--background)",
        borderColor: "transparent",
        borderWidth: "0px",
        button: {
          foreground: "var(--theme--foreground-subdued)",
          foregroundHover: "#fff",
          foregroundActive: "var(--theme--foreground-accent)",
          background: "transparent",
          backgroundHover: "transparent",
          backgroundActive: "#21262e"
        }
      },
      list: {
        icon: {
          foreground: "var(--theme--primary)",
          foregroundHover: "var(--theme--navigation--list--icon--foreground)",
          foregroundActive: "var(--theme--navigation--list--icon--foreground)"
        },
        foreground: "var(--theme--foreground-accent)",
        foregroundHover: "var(--theme--navigation--list--foreground)",
        foregroundActive: "var(--theme--navigation--list--foreground)",
        background: "transparent",
        backgroundHover: "#30363d",
        backgroundActive: "#30363d",
        fontFamily: "var(--theme--fonts--sans--font-family)",
        divider: {
          borderColor: "#30363d",
          borderWidth: "var(--theme--border-width)"
        }
      }
    },
    header: {
      background: "var(--theme--background)",
      borderColor: "transparent",
      borderWidth: "0px",
      boxShadow: "0 4px 7px -4px black",
      headline: {
        foreground: "var(--theme--foreground-subdued)",
        fontFamily: "var(--theme--fonts--sans--font-family)"
      },
      title: {
        foreground: "var(--theme--foreground-accent)",
        fontFamily: "var(--theme--fonts--display--font-family)",
        fontWeight: "var(--theme--fonts--display--font-weight)"
      }
    },
    form: {
      columnGap: "32px",
      rowGap: "40px",
      field: {
        label: {
          foreground: "var(--theme--foreground-accent)",
          fontFamily: "var(--theme--fonts--sans--font-family)",
          fontWeight: "600"
        },
        input: {
          background: "var(--theme--background)",
          backgroundSubdued: "var(--theme--background-subdued)",
          foreground: "var(--theme--foreground)",
          foregroundSubdued: "var(--theme--foreground-subdued)",
          borderColor: "#21262e",
          borderColorHover: "#30363d",
          borderColorFocus: "var(--theme--primary)",
          boxShadow: "none",
          boxShadowHover: "none",
          boxShadowFocus: "0 0 16px -8px var(--theme--primary)",
          height: "60px",
          padding: "16px"
        }
      }
    },
    sidebar: {
      background: "#21262e",
      foreground: "var(--theme--foreground-subdued)",
      fontFamily: "var(--theme--fonts--sans--font-family)",
      borderColor: "transparent",
      borderWidth: "0px",
      section: {
        toggle: {
          icon: {
            foreground: "var(--theme--foreground-accent)",
            foregroundHover: "var(--theme--sidebar--section--toggle--icon--foreground)",
            foregroundActive: "var(--theme--sidebar--section--toggle--icon--foreground)"
          },
          foreground: "var(--theme--foreground-accent)",
          foregroundHover: "var(--theme--sidebar--section--toggle--foreground)",
          foregroundActive: "var(--theme--sidebar--section--toggle--foreground)",
          background: "#30363d",
          backgroundHover: "var(--theme--sidebar--section--toggle--background)",
          backgroundActive: "var(--theme--sidebar--section--toggle--background)",
          fontFamily: "var(--theme--fonts--sans--font-family)",
          borderColor: "transparent",
          borderWidth: "0px"
        },
        form: {
          columnGap: "var(--theme--form--column-gap)",
          rowGap: "var(--theme--form--row-gap)",
          label: {
            foreground: "var(--theme--form--field--label--foreground)",
            fontFamily: "var(--theme--form--field--label--font-family)"
          },
          field: {
            input: {
              background: "var(--theme--form--field--input--background)",
              foreground: "var(--theme--form--field--input--foreground)",
              foregroundSubdued: "var(--theme--form--field--input--foreground-subdued)",
              borderColor: "var(--theme--form--field--input--border-color)",
              borderColorHover: "var(--theme--form--field--input--border-color-hover)",
              borderColorFocus: "var(--theme--form--field--input--border-color-focus)",
              boxShadow: "var(--theme--form--field--input--box-shadow)",
              boxShadowHover: "var(--theme--form--field--input--box-shadow-hover)",
              boxShadowFocus: "var(--theme--form--field--input--box-shadow-focus)",
              height: "52px",
              padding: "12px"
            }
          }
        }
      }
    },
    public: {
      background: "var(--theme--background)",
      foreground: "var(--theme--foreground)",
      foregroundAccent: "var(--theme--foreground-accent)",
      art: {
        background: "#0e1c2f",
        primary: "var(--theme--primary)",
        secondary: "var(--theme--secondary)",
        speed: "1"
      },
      form: {
        columnGap: "var(--theme--form--column-gap)",
        rowGap: "var(--theme--form--row-gap)",
        field: {
          label: {
            foreground: "var(--theme--form--field--label--foreground)",
            fontFamily: "var(--theme--form--field--label--font-family)"
          },
          input: {
            background: "var(--theme--form--field--input--background)",
            foreground: "var(--theme--form--field--input--foreground)",
            foregroundSubdued: "var(--theme--form--field--input--foreground-subdued)",
            borderColor: "var(--theme--form--field--input--border-color)",
            borderColorHover: "var(--theme--form--field--input--border-color-hover)",
            borderColorFocus: "var(--theme--form--field--input--border-color-focus)",
            boxShadow: "var(--theme--form--field--input--box-shadow)",
            boxShadowHover: "var(--theme--form--field--input--box-shadow-hover)",
            boxShadowFocus: "var(--theme--form--field--input--box-shadow-focus)",
            height: "var(--theme--form--field--input--height)",
            padding: "var(--theme--form--field--input--padding)"
          }
        }
      }
    },
    popover: {
      menu: {
        background: "#161b22",
        borderRadius: "var(--theme--border-radius)",
        boxShadow: "0px 0px 6px 0px black"
      }
    },
    banner: {
      background: "#0e1c2f",
      padding: "40px",
      borderRadius: "var(--theme--border-radius)",
      avatar: {
        borderRadius: "50%",
        foreground: "var(--theme--primary)",
        background: "#ffffff"
      },
      headline: {
        foreground: "#ffffff",
        fontFamily: "var(--theme--fonts--sans--font-family)",
        fontWeight: "var(--theme--fonts--sans--font-weight)"
      },
      title: {
        foreground: "#ffffff",
        fontFamily: "var(--theme--fonts--display--font-family)",
        fontWeight: "var(--theme--fonts--display--font-weight)"
      },
      subtitle: {
        foreground: "#a2b5cd",
        fontFamily: "var(--theme--fonts--monospace--font-family)",
        fontWeight: "var(--theme--fonts--monospace--font-weight)"
      },
      art: {
        foreground: "#2e3a4d"
      }
    }
  }
}), fg = An({
  id: "Directus Color Match",
  name: "$t:theme_directus_colormatch",
  appearance: "light",
  rules: {
    background: "color-mix(in srgb, #FFFFFF, var(--theme--primary) 7%)",
    backgroundAccent: "color-mix(in srgb, #FFFFFF, var(--theme--primary) 20%)",
    backgroundNormal: "color-mix(in srgb, #FFFFFF, var(--theme--primary) 15%)",
    backgroundSubdued: "color-mix(in srgb, #FFFFFF, var(--theme--primary) 10%)",
    borderColor: "color-mix(in srgb, #FFFFFF, var(--theme--primary) 20%)",
    borderColorAccent: "color-mix(in srgb, #FFFFFF, var(--theme--primary) 40%)",
    borderColorSubdued: "color-mix(in srgb, #FFFFFF, var(--theme--primary) 15%)",
    borderRadius: "12px",
    borderWidth: "1px",
    foreground: "color-mix(in srgb, #000000, var(--theme--primary) 70%)",
    foregroundAccent: "color-mix(in srgb, #000000, var(--theme--primary) 50%)",
    foregroundSubdued: "color-mix(in srgb, #FFFFFF, var(--theme--primary) 60%)",
    fonts: {
      display: {
        fontFamily: '"Montserrat", system-ui',
        fontWeight: "400"
      }
    },
    form: {
      field: {
        input: { background: "#FFFFFF", backgroundSubdued: "color-mix(in srgb, #FFFFFF, var(--theme--primary) 13%)" }
      }
    },
    navigation: {
      background: "#FFFFFF",
      backgroundAccent: "var(--theme--background)",
      borderWidth: "var(--theme--border-width)",
      borderColor: "var(--theme--border-color-subdued)",
      modules: {
        background: "color-mix(in srgb, #000000, var(--theme--primary) 90%)",
        button: {
          backgroundActive: "#FFFFFF",
          foreground: "color-mix(in srgb, #FFFFFF, var(--theme--primary) 20%)",
          foregroundActive: "var(--theme--primary)"
        }
      },
      project: { borderWidth: "1px", background: "#FFFFFF", borderColor: "var(--theme--border-color-subdued)" },
      list: {
        divider: { borderColor: "var(--theme--border-color-subdued)" },
        icon: { foreground: "var(--theme--foreground)" },
        foreground: "var(--theme--foreground)",
        foregroundHover: "var(--theme--foreground)",
        foregroundActive: "var(--theme--foreground)"
      }
    },
    header: {
      background: "#FFFFFF",
      borderWidth: "1px",
      borderColor: "var(--theme--border-color-subdued)",
      boxShadow: "0 4px 7px -4px rgba(0,102,102, 0.2)"
    },
    sidebar: {
      background: "#FFFFFF",
      borderWidth: "1px",
      borderColor: "var(--theme--border-color-subdued)",
      section: {
        toggle: {
          borderColor: "var(--theme--border-color-subdued)",
          borderWidth: "1px",
          background: "#FFFFFF",
          foreground: "var(--theme--foreground)",
          foregroundHover: "var(--theme--foreground)",
          foregroundActive: "var(--theme--foreground-accent)",
          icon: {
            foreground: "var(--theme--foreground)",
            foregroundHover: "var(--theme--foreground)",
            foregroundActive: "var(--theme--foreground-accent)"
          }
        }
      }
    },
    public: {
      art: {
        background: "color-mix(in srgb, #FFFFFF, var(--theme--primary) 10%)",
        primary: "color-mix(in srgb, #FFFFFF, var(--theme--primary) 60%)",
        secondary: "color-mix(in srgb, #FFFFFF, var(--theme--secondary) 70%)"
      },
      background: "#FFFFFF"
    }
  }
}), jt = An({
  id: "Directus Default",
  name: "$t:theme_directus_default",
  appearance: "light",
  rules: {
    borderRadius: "6px",
    borderWidth: "2px",
    foreground: "#4f5464",
    foregroundAccent: "#172940",
    foregroundSubdued: "#a2b5cd",
    background: "#fff",
    backgroundNormal: "#f0f4f9",
    backgroundAccent: "#e4eaf1",
    backgroundSubdued: "#f7fafc",
    borderColor: "#e4eaf1",
    borderColorAccent: "#d3dae4",
    borderColorSubdued: "#f0f4f9",
    primary: "var(--project-color)",
    primaryBackground: "color-mix(in srgb, var(--theme--background), var(--theme--primary) 10%)",
    primarySubdued: "color-mix(in srgb, var(--theme--background), var(--theme--primary) 50%)",
    primaryAccent: "color-mix(in srgb, var(--theme--primary), #2e3c43 25%)",
    secondary: "#ff99dd",
    secondaryBackground: "color-mix(in srgb, var(--theme--background), var(--theme--secondary) 10%)",
    secondarySubdued: "color-mix(in srgb, var(--theme--background), var(--theme--secondary) 50%)",
    secondaryAccent: "color-mix(in srgb, var(--theme--secondary), #2e3c43 25%)",
    success: "#2ecda7",
    successBackground: "color-mix(in srgb, var(--theme--background), var(--theme--success) 10%)",
    successSubdued: "color-mix(in srgb, var(--theme--background), var(--theme--success) 50%)",
    successAccent: "color-mix(in srgb, var(--theme--success), #2e3c43 25%)",
    warning: "#ffa439",
    warningBackground: "color-mix(in srgb, var(--theme--background), var(--theme--warning) 10%)",
    warningSubdued: "color-mix(in srgb, var(--theme--background), var(--theme--warning) 50%)",
    warningAccent: "color-mix(in srgb, var(--theme--warning), #2e3c43 25%)",
    danger: "#e35169",
    dangerBackground: "color-mix(in srgb, var(--theme--background), var(--theme--danger) 10%)",
    dangerSubdued: "color-mix(in srgb, var(--theme--background), var(--theme--danger) 50%)",
    dangerAccent: "color-mix(in srgb, var(--theme--danger), #2e3c43 25%)",
    fonts: {
      display: {
        fontFamily: '"Inter", system-ui',
        fontWeight: "700"
      },
      sans: {
        fontFamily: '"Inter", system-ui',
        fontWeight: "500"
      },
      serif: {
        fontFamily: '"Merriweather", serif',
        fontWeight: "500"
      },
      monospace: {
        fontFamily: '"Fira Mono", monospace',
        fontWeight: "500"
      }
    },
    navigation: {
      background: "var(--theme--background-normal)",
      backgroundAccent: "var(--theme--background-accent)",
      borderColor: "transparent",
      borderWidth: "0px",
      project: {
        borderColor: "transparent",
        borderWidth: "0px",
        background: "var(--theme--navigation--background-accent)",
        foreground: "var(--theme--foreground-accent)",
        fontFamily: "var(--theme--font-family-sans-serif)"
      },
      modules: {
        background: "#0e1c2f",
        borderColor: "transparent",
        borderWidth: "0px",
        button: {
          foreground: "#8196b1",
          foregroundHover: "#fff",
          foregroundActive: "var(--theme--foreground-accent)",
          background: "transparent",
          backgroundHover: "transparent",
          backgroundActive: "var(--theme--background-normal)"
        }
      },
      list: {
        icon: {
          foreground: "var(--theme--primary)",
          foregroundHover: "var(--theme--navigation--list--icon--foreground)",
          foregroundActive: "var(--theme--navigation--list--icon--foreground)"
        },
        foreground: "var(--theme--foreground-accent)",
        foregroundHover: "var(--theme--navigation--list--foreground)",
        foregroundActive: "var(--theme--navigation--list--foreground)",
        background: "transparent",
        backgroundHover: "var(--theme--navigation--background-accent)",
        backgroundActive: "var(--theme--navigation--background-accent)",
        fontFamily: "var(--theme--fonts--sans--font-family)",
        divider: {
          borderColor: "var(--theme--border-color-accent)",
          borderWidth: "var(--theme--border-width)"
        }
      }
    },
    header: {
      background: "var(--theme--background)",
      borderColor: "transparent",
      borderWidth: "0px",
      boxShadow: "0 4px 7px -4px rgb(0 0 0 / 0.2)",
      headline: {
        foreground: "var(--theme--foreground-subdued)",
        fontFamily: "var(--theme--fonts--sans--font-family)"
      },
      title: {
        foreground: "var(--theme--foreground-accent)",
        fontFamily: "var(--theme--fonts--display--font-family)",
        fontWeight: "var(--theme--fonts--display--font-weight)"
      }
    },
    form: {
      columnGap: "32px",
      rowGap: "40px",
      field: {
        label: {
          foreground: "var(--theme--foreground-accent)",
          fontFamily: "var(--theme--fonts--sans--font-family)",
          fontWeight: "600"
        },
        input: {
          background: "var(--theme--background)",
          backgroundSubdued: "var(--theme--background-subdued)",
          foreground: "var(--theme--foreground)",
          foregroundSubdued: "var(--theme--foreground-subdued)",
          borderColor: "var(--theme--border-color)",
          borderColorHover: "var(--theme--border-color-accent)",
          borderColorFocus: "var(--theme--primary)",
          boxShadow: "none",
          boxShadowHover: "none",
          boxShadowFocus: "0 0 16px -8px var(--theme--primary)",
          height: "60px",
          padding: "16px"
        }
      }
    },
    sidebar: {
      background: "var(--theme--background-normal)",
      foreground: "var(--theme--foreground-subdued)",
      fontFamily: "var(--theme--fonts--sans--font-family)",
      borderColor: "transparent",
      borderWidth: "0px",
      section: {
        toggle: {
          icon: {
            foreground: "var(--theme--foreground-accent)",
            foregroundHover: "var(--theme--sidebar--section--toggle--icon--foreground)",
            foregroundActive: "var(--theme--sidebar--section--toggle--icon--foreground)"
          },
          foreground: "var(--theme--foreground-accent)",
          foregroundHover: "var(--theme--sidebar--section--toggle--foreground)",
          foregroundActive: "var(--theme--sidebar--section--toggle--foreground)",
          background: "var(--theme--background-accent)",
          backgroundHover: "var(--theme--sidebar--section--toggle--background)",
          backgroundActive: "var(--theme--sidebar--section--toggle--background)",
          fontFamily: "var(--theme--fonts--sans--font-family)",
          borderColor: "transparent",
          borderWidth: "0px"
        },
        form: {
          columnGap: "var(--theme--form--column-gap)",
          rowGap: "var(--theme--form--row-gap)",
          label: {
            foreground: "var(--theme--form--field--label--foreground)",
            fontFamily: "var(--theme--form--field--label--font-family)"
          },
          field: {
            input: {
              background: "var(--theme--form--field--input--background)",
              foreground: "var(--theme--form--field--input--foreground)",
              foregroundSubdued: "var(--theme--form--field--input--foreground-subdued)",
              borderColor: "var(--theme--form--field--input--border-color)",
              borderColorHover: "var(--theme--form--field--input--border-color-hover)",
              borderColorFocus: "var(--theme--form--field--input--border-color-focus)",
              boxShadow: "var(--theme--form--field--input--box-shadow)",
              boxShadowHover: "var(--theme--form--field--input--box-shadow-hover)",
              boxShadowFocus: "var(--theme--form--field--input--box-shadow-focus)",
              height: "52px",
              padding: "12px"
            }
          }
        }
      }
    },
    public: {
      background: "var(--theme--background)",
      foreground: "var(--theme--foreground)",
      foregroundAccent: "var(--theme--foreground-accent)",
      art: {
        background: "#0e1c2f",
        primary: "var(--theme--primary)",
        secondary: "var(--theme--secondary)",
        speed: "1"
      },
      form: {
        columnGap: "var(--theme--form--column-gap)",
        rowGap: "var(--theme--form--row-gap)",
        label: {
          foreground: "var(--theme--form--field--label--foreground)",
          fontFamily: "var(--theme--form--field--label--font-family)"
        },
        field: {
          input: {
            background: "var(--theme--form--field--input--background)",
            foreground: "var(--theme--form--field--input--foreground)",
            foregroundSubdued: "var(--theme--form--field--input--foreground-subdued)",
            borderColor: "var(--theme--form--field--input--border-color)",
            borderColorHover: "var(--theme--form--field--input--border-color-hover)",
            borderColorFocus: "var(--theme--form--field--input--border-color-focus)",
            boxShadow: "var(--theme--form--field--input--box-shadow)",
            boxShadowHover: "var(--theme--form--field--input--box-shadow-hover)",
            boxShadowFocus: "var(--theme--form--field--input--box-shadow-focus)",
            height: "var(--theme--form--field--input--height)",
            padding: "var(--theme--form--field--input--padding)"
          }
        }
      }
    },
    popover: {
      menu: {
        background: "#fafcfd",
        borderRadius: "var(--theme--border-radius)",
        boxShadow: "0px 0px 6px 0px rgb(23, 41, 64, 0.2), 0px 0px 12px 2px rgb(23, 41, 64, 0.05)"
      }
    },
    banner: {
      background: "#0e1c2f",
      padding: "40px",
      borderRadius: "var(--theme--border-radius)",
      avatar: {
        borderRadius: "50%",
        foreground: "var(--theme--primary)",
        background: "#ffffff"
      },
      headline: {
        foreground: "#ffffff",
        fontFamily: "var(--theme--fonts--sans--font-family)",
        fontWeight: "var(--theme--fonts--sans--font-weight)"
      },
      title: {
        foreground: "#ffffff",
        fontFamily: "var(--theme--fonts--display--font-family)",
        fontWeight: "var(--theme--fonts--display--font-weight)"
      },
      subtitle: {
        foreground: "#a2b5cd",
        fontFamily: "var(--theme--fonts--monospace--font-family)",
        fontWeight: "var(--theme--fonts--monospace--font-weight)"
      },
      art: {
        foreground: "#2e3a4d"
      }
    }
  }
}), dg = An({
  id: "Directus Minimal",
  name: "$t:theme_directus_minimal",
  appearance: "light",
  rules: {
    borderWidth: "1px",
    backgroundPage: "color-mix(in srgb, #FFFFFF, var(--theme--primary) 7%)",
    navigation: {
      background: "#FFFFFF",
      modules: {
        background: "#FFFFFF",
        button: {
          backgroundActive: "#F1F5F9",
          foreground: "var(--theme--foreground)",
          foregroundHover: "var(--theme--primary)",
          foregroundActive: "var(--theme--primary)",
          backgroundHover: "#F1F5F9",
          background: "#FFFFFF"
        },
        borderWidth: "1px",
        borderColor: "var(--theme--border-color)"
      },
      project: {
        borderWidth: "1px",
        background: "#FFFFFF",
        borderColor: "var(--theme--border-color)"
      },
      list: {
        icon: {
          foreground: "#0F172A"
        },
        divider: {
          borderColor: "var(--theme--border-color)"
        }
      },
      borderWidth: "1px",
      backgroundAccent: "#F1F5F9",
      borderColor: "var(--theme--border-color)"
    },
    header: {
      background: "#FFFFFF",
      borderWidth: "1px",
      borderColor: "var(--theme--border-color)",
      boxShadow: "0 4px 7px -4px rgba(0,102,102, 0.1)"
    },
    backgroundAccent: "#E2E8F0",
    backgroundSubdued: "#F8FAFC",
    background: "#FFFFFF",
    foreground: "#1E293B",
    foregroundAccent: "#0F172A",
    foregroundSubdued: "#94A3B8",
    borderRadius: "4px",
    borderColor: "#E2E8F0",
    borderColorAccent: "#CBD5E1",
    borderColorSubdued: "#F1F5F9",
    form: {
      rowGap: "32px",
      field: {
        input: {
          background: "#FFFFFF",
          backgroundSubdued: "#F8FAFC",
          boxShadowFocus: "none",
          height: "52px"
        }
      }
    },
    sidebar: {
      background: "#FFFFFF",
      borderWidth: "1px",
      borderColor: "var(--theme--border-color)",
      section: {
        toggle: {
          borderColor: "var(--theme--border-color)",
          borderWidth: "1px",
          background: "#FFFFFF",
          foreground: "var(--theme--foreground-subdued)",
          foregroundHover: "var(--theme--foreground)",
          foregroundActive: "var(--theme--foreground-accent)",
          icon: {
            foreground: "var(--theme--foreground)",
            foregroundHover: "var(--theme--foreground)",
            foregroundActive: "var(--theme--foreground-accent)"
          }
        },
        form: {
          field: {
            input: {
              height: "42px"
            }
          }
        }
      }
    },
    public: {
      art: {
        background: "color-mix(in srgb, #FFFFFF, var(--project-color) 10%)",
        primary: "color-mix(in srgb, #FFFFFF, var(--project-color) 70%)",
        secondary: "color-mix(in srgb, #FFFFFF, var(--project-color) 40%)"
      },
      background: "#FFFFFF"
    },
    backgroundNormal: "#F1F5F9",
    secondary: "#64748B",
    primary: "#0F172A",
    primaryBackground: "#F1F5F9",
    primarySubdued: "#F8FAFC",
    primaryAccent: "#E2E8F0",
    secondaryAccent: "#E2E8F0",
    secondaryBackground: "#F1F5F9",
    secondarySubdued: "#F8FAFC",
    fonts: {
      display: {
        fontFamily: "system-ui"
      }
    }
  }
}), mg = [Zt], lg = [jt, dg, fg], pg = Vi(" Themes", () => {
  const e = Ei({ light: lg, dark: mg });
  return { themes: e, registerTheme: (n) => {
    n.appearance === "light" ? e.light.push(n) : e.dark.push(n);
  } };
}), hg = (e, r, n, t, o) => {
  const { themes: u } = Hi(pg());
  return { theme: Ce(() => {
    const a = ee(e) ? ee(n) : ee(r), l = ee(e) ? Zt : jt, p = ee(e) ? ee(o) : ee(t), b = ee(u)[ee(e) ? "dark" : "light"].find((y) => y.id === a);
    return b ? p ? Tn({}, l, b, { rules: p }) : Tn(l, b) : (a && a !== l.id && console.warn(`Theme "${a}" doesn't exist.`), p ? Tn({}, l, { rules: p }) : l);
  }) };
}, gg = (e) => {
  const r = qi(e, { delimiter: "--" }), n = (t) => `--theme--${Gi(t, { separator: "-" })}`;
  return Ki(r, (t, o) => n(o));
}, yg = /* @__PURE__ */ Ci({
  __name: "theme-provider",
  props: {
    darkMode: { type: Boolean },
    themeLight: { default: jt.name },
    themeLightOverrides: { default: () => ({}) },
    themeDark: { default: Zt.name },
    themeDarkOverrides: { default: () => ({}) }
  },
  setup(e) {
    const r = e, { darkMode: n, themeLight: t, themeDark: o, themeLightOverrides: u, themeDarkOverrides: c } = Ni(r), { theme: a } = hg(n, t, o, u, c), l = Ce(() => gg(ee(a).rules)), { googleFonts: p } = sg(a);
    Bi({
      link: Ce(() => {
        let y = "";
        if (p.value.length > 0) {
          const O = p.value.join("&family=");
          y += `https://fonts.googleapis.com/css2?family=${O}`, y += `
`;
        }
        return y ? [
          {
            rel: "stylesheet",
            href: y
          }
        ] : [];
      })
    });
    const b = Ce(() => `:root {${Object.entries(ee(l)).map(([O, I]) => `${O}: ${I};`).join(" ")}}`);
    return (y, O) => (Di(), Ui(Li, { to: "#theme" }, [
      Mi(Wi(b.value), 1)
    ]));
  }
});
export {
  yg as ThemeProvider,
  An as defineTheme,
  gg as rulesToCssVars,
  sg as useFonts,
  hg as useTheme,
  pg as useThemeStore
};
